---
title: "Re-analysis of NicheNet dataset"
---
  
```{r setup, include=F}
library(colorspace)
library(scales)
source("~/Dropbox/GDB/line2user.R")
knitr::opts_chunk$set(
  echo=F,
  message=F,
  warning=F
)
```


# NicheNet calculation sanity check

```{r load_raw}
setwd("~/Dropbox/GDB_archive/CMapCorr_files/GSE47542_RAW/")

temp_TNF_p <- c("GSM1151945_Agilent_251485053931_S01_GE2_107_Sep09_1_4.txt.gz",
                "GSM1151946_Agilent_251485053932_S01_GE2_107_Sep09_1_4.txt.gz",
                "GSM1151950_Agilent_251485054074_S01_GE2_107_Sep09_1_4.txt.gz")
temp_TNF_m <- c("GSM1151933_Agilent_251485053931_S01_GE2_107_Sep09_1_2.txt.gz",
                "GSM1151934_Agilent_251485053932_S01_GE2_107_Sep09_1_2.txt.gz",
                "GSM1151938_Agilent_251485054074_S01_GE2_107_Sep09_1_2.txt.gz")
temp_CTRL_p <- c("GSM1151939_Agilent_251485053931_S01_GE2_107_Sep09_1_3.txt.gz",
                 "GSM1151940_Agilent_251485053932_S01_GE2_107_Sep09_1_3.txt.gz",
                 "GSM1151947_Agilent_251485054010_S01_GE2_107_Sep09_1_4.txt.gz")
temp_CTRL_m <- c("GSM1151928_Agilent_251485053932_S01_GE2_107_Sep09_1_1.txt.gz",
                 "GSM1151935_Agilent_251485054010_S01_GE2_107_Sep09_1_2.txt.gz",
                 "GSM1151951_Agilent_251485053931_S01_GE2_107_Sep09_1_1.txt.gz")

TNFp <- sapply(temp_TNF_p,function(X) {
  DS <- read.table(X,sep="\t",header=T,
                   skip=which(grepl("^FEAT",scan(X,"character",sep="\n"))) - 1)
  return(DS[DS$SystematicName == "NM_000201","LogRatio"])
},USE.NAMES=F)

TNFm <- sapply(temp_TNF_m,function(X) {
  DS <- read.table(X,sep="\t",header=T,
                   skip=which(grepl("^FEAT",scan(X,"character",sep="\n"))) - 1)
  return(DS[DS$SystematicName == "NM_000201","LogRatio"])
},USE.NAMES=F)

CTRLp <- sapply(temp_CTRL_p,function(X) {
  DS <- read.table(X,sep="\t",header=T,
                   skip=which(grepl("^FEAT",scan(X,"character",sep="\n"))) - 1)
  return(DS[DS$SystematicName == "NM_000201","LogRatio"])
},USE.NAMES=F)

CTRLm <- sapply(temp_CTRL_m,function(X) {
  DS <- read.table(X,sep="\t",header=T,
                   skip=which(grepl("^FEAT",scan(X,"character",sep="\n"))) - 1)
  return(DS[DS$SystematicName == "NM_000201","LogRatio"])
},USE.NAMES=F)

rm(list=grep("^temp",ls(),value=T))
```

```{r importNN}
nn_db <- readRDS(url("https://zenodo.org/record/3260758/files/expression_settings.rds"))
nn_db <- nn_db[sapply(sapply(nn_db,"[[","from"),length) == 1]

diff_M <- nn_db$cifola_Tnf$diffexp[nn_db$cifola_Tnf$diffexp$gene == "ICAM1",c("lfc","qval")]
diff_P <- nn_db$cifola_Tnf_plastic$diffexp[nn_db$cifola_Tnf_plastic$diffexp$gene == "ICAM1",c("lfc","qval")]
```

```{r plot_raw,fig.height=4,fig.width=6}
par(mar=c(3,3,3,1),mgp=2:0)
boxplot(list(TNFp=as.vector(TNFp),
             CTRLp=as.vector(CTRLp),
             TNFm=as.vector(TNFm),
             CTRLm=as.vector(CTRLm)),
        main="ICAM1 expression",
        ylab="Log-ratio vs reference RNA")

text(1.5,par("usr")[3] + (par("usr")[4] - par("usr")[3]) / 2,
     paste("NicheNet",
           paste0("logFC = ",signif(diff_P$lfc,2)),
           paste0("FDR = ",signif(diff_P$qval,2)),
           sep="\n"))

text(3.5,par("usr")[3] + (par("usr")[4] - par("usr")[3]) / 2,
     paste("NicheNet",
           paste0("logFC = ",signif(diff_M$lfc,2)),
           paste0("FDR = ",signif(diff_M$qval,2)),
           sep="\n"))

```

Looks like the sign of the logFC is flipped in the NicheNet calculations.  Multiplying logFC values from the NicheNet dataset by -1 for the remainder of this analysis.

```{r cleanup_sanity_check}
rm(list=ls())
source("~/Dropbox/GDB/line2user.R")
```

****

# NicheNet data summary

```{r AvC_load_data}
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_ALLvCT_dat.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_ALLvCT_corr.RData")
# load("~/Dropbox/GDB_archive/CMapCorr_files/NN_ALLvCT_DEovlp.RData")
```


```{r DSsummary,fig.height=6,fig.width=7}
temp_DScount <- table(do.call(rbind,DSinfo)[,c("cell_type","ligand")])
temp_DScount <- temp_DScount[,rev(colnames(temp_DScount))]

temp_ligcount <- apply(temp_DScount,1,function(X) sum(X > 0))
temp_ctcount <- apply(temp_DScount,2,function(X) sum(X > 0))

layout(matrix(c(2,1,0,3),2),widths=c(4,1),heights=c(1,4))
par(mar=c(7,4,0,0),mgp=2:0)
image(x=1:nrow(temp_DScount),y=1:ncol(temp_DScount),z=temp_DScount,
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(max(temp_DScount) + 1,palette="purple-yellow",rev=T))
abline(h=seq(1.5,ncol(temp_DScount)),col="grey80",lwd=0.5)
abline(v=seq(1.5,nrow(temp_DScount)),col="grey80",lwd=0.5)
text(1:nrow(temp_DScount),rep(line2user(0.5,1),nrow(temp_DScount)),
     rownames(temp_DScount),srt=45,adj=c(1,0.5),xpd=NA,cex=0.9)
mtext("Cell Type",side=1,line=6,font=2)
mtext(colnames(temp_DScount),at=1:ncol(temp_DScount),
      side=2,las=2,line=0.1,cex=0.9)
mtext("Ligand",side=2,line=3,font=2)
text(which(temp_DScount > 0,arr.ind=T),
     labels=temp_DScount[temp_DScount > 0],
     col="lightgoldenrodyellow",cex=0.9)

par(mar=c(0.1,4,1,0))
plot(temp_ligcount,type="h",lwd=20,lend=1,col="grey80",
     bty="n",xaxt="n",xaxs="i",yaxs="i",
     xlim=c(0.5,length(temp_ligcount) + 0.5),
     ylim=c(0.5,max(temp_ligcount) + 0.5),
     ylab=paste("Total","unique ligands",sep="\n"))

par(mar=c(7,0.1,0,1),mgp=c(3,1,0))
plot(NA,NA,xlim=c(0.5,max(temp_ctcount) + 0.5),ylim=c(0.5,length(temp_ctcount) + 0.5),
     yaxt="n",yaxs="i",xaxs="i",bty="n",xlab=paste("Total unique","cell types",sep="\n"))
temp <- sapply(seq_along(temp_ctcount),function(X)
  points(c(0,temp_ctcount[X]),c(X,X),type="l",lwd=15,lend=1,col="grey80"))
# plot(temp_ctcount,type="h",,
#      xaxt="n",xaxs="i",bty="n",ylab=paste("Total","unique ligands",sep="\n"),
#      xlim=c(0.5,length(temp_ligcount) + 0.5))

```

`r length(temp_ctcount)` ligands were assayed in multiple datasets, `r sum(temp_ctcount > 1)` of which occurred across more than one cell type.

```{r DSsummary_cleanup}
rm(list=grep("^temp",ls(),value=T))
```


****
# Comparing across vs within cell type per ligand treatment 

Previous analysis (see below) included a "within lab" category that was unclear, so it has been removed.  
All data (including time-series) is included in this analysis.  NicheNet calculation of time-series data seems to result in all positive logFC values, so that makes some things a little screwy.  In those cases, time-series data was only compared to other time-series data.  

## Correlation

```{r AvC_corr_plot,fig.height=4,fig.width=8}
temp_order <- names(sort(sapply(unique(c(names(corr_all),names(corr_ct))),
                                function(X) mean(c(corr_all[[X]],corr_ct[[X]])))))
lig_col <- qualitative_hcl(length(nn_ligands),palette="dark3",alpha=0.5)
temp_colnames <- c(temp_order,nn_ligands[!nn_ligands %in% temp_order])
names(lig_col) <- temp_colnames[unlist(sapply(seq(1,ceiling(length(nn_ligands)/2)),
                                              function(X) c(X,X + ceiling(length(nn_ligands)/2)),
                                              simplify=F))]
temp_corr <- list(ALL=corr_all[temp_order[temp_order %in% names(corr_all)]],
               CT=corr_ct[temp_order[temp_order %in% names(corr_ct)]])
temp_lig <- sapply(strsplit(names(unlist(temp_corr,recursive=F)),".",fixed=T),"[[",2)


layout(rbind(1:2),widths=c(4,1))

par(mar=c(5,3,1,0.5),mgp=2:0)
plot(NA,NA,xlim=c(1,length(unlist(temp_corr,recursive=F))),ylim=range(temp_corr),
     xaxt="n",xlab=NA,ylab="Spearman correlation of log(fold-change)")
abline(v=sapply(temp_corr,length)[1] + 0.5,col="grey70")
abline(h=0,lty=2,col="grey50")
temp_x <- unlist(mapply(rep,x=seq_along(unlist(temp_corr,recursive=F)),
                        times=sapply(unlist(temp_corr,recursive=F),length)))
points(x=jitter(temp_x,amount=0.2),y=unlist(temp_corr),
       pch=19,col=lig_col[temp_lig][temp_x])
mtext(temp_lig,side=1,line=0.1,at=seq_along(temp_lig),las=3,col=alpha(lig_col[temp_lig],1))

segments(x0=c(0,sapply(temp_corr,length)[1]) + 0.75,
         x1=cumsum(sapply(temp_corr,length)) + 0.25,
         y0=line2user(4,1),y1=line2user(4,1),xpd=NA,lwd=2)
segments(x0=c(c(0,sapply(temp_corr,length)[1]) + 0.75,
              cumsum(sapply(temp_corr,length)) + 0.25),
         x1=c(c(0,sapply(temp_corr,length)[1]) + 0.75,
              cumsum(sapply(temp_corr,length)) + 0.25),
         y0=line2user(4,1),y1=line2user(3.8,1),xpd=NA,lwd=2)
mtext(c("Pairwise across all cell types","Pairwise within same cell type"),
      at=0.5 + cumsum(sapply(temp_corr,length)) - sapply(temp_corr,length) / 2,
      side=1,line=4)


par(mar=c(5,0,1,3))
boxplot(list(ALL=unlist(corr_all),CT=unlist(corr_ct)),
        xaxt="n",xlab=NA,yaxt="n",ylab=NA)
axis(4)
mtext("Spearman correlation of log(fold-change)",side=4,line=2)
mtext(c("Across all","Cell type"),side=1,line=0.1,at=1:2,las=3)

mtext(paste0("p=",signif(wilcox.test(unlist(corr_all),unlist(corr_ct))$p.value,2)),
      side=1,line=-0.1,at=1.5,las=3,adj=0,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```

## Differential gene expression

Differentially expressed genes were determined by |logFC| >= 1 and FDR <= 0.1 (similar to Browaeys *et al.*, 2020).  This figure shows number of DE genes overlapping across all datasets testing the same ligand vs. datasets testing the same ligand in the same cell type vs. datasets from the same accession number (and same ligand / cell type).  P-values for overlap (y-axis) were determined after determining probability of overlapping DE genes by chance through bootstrapping.

```{r AvC_diffexp_boxplot_v2,fig.height=8,fig.width=9}
par(mfrow=c(2,6),mgp=2:0)
for (charLFC in c("2","4")) {
  for (charFDR in c("1","01")) {
    temp_ds <- load(paste0("~/Dropbox/GDB_archive/CMapCorr_files/NN_ALLvCT_DEovlp_",
                        "LFC",charLFC,"FDR",charFDR,".RData"))
    LFC <- log2(as.integer(charLFC))
    FDR <- as.numeric(paste0("0.",charFDR))
    temp_y <- lapply(c(ALL="P_all",CT="P_ct"),function(X) -log10(unlist(get(X))))
    temp_x <- list(ALL=sapply(de_all,length),
                   CT=sapply(unlist(de_ct,F,T),length))
    temp_CTnames <- names(temp_y$CT)
    names(temp_x$CT) <- names(temp_y$CT) <- sapply(strsplit(temp_CTnames,".",fixed=T),"[[",1)
      
    par(mar=c(5,4,2,0.5))
    boxplot(temp_y,ylim=range(temp_y),outline=T,xaxt="n",yaxt="n",ylab=NA)
    text(1.5,line2user(-0.2,3),srt=90,adj=1,cex=1.2,
         labels=paste("p =",signif(wilcox.test(unlist(P_all),unlist(P_ct))$p.value,2)))
    axis(2,cex.axis=1.2)
    mtext("-log10(P # DE overlapping by chance)",side=2,line=2.5)
    text(1:2,line2user(0.5,1),xpd=NA,srt=45,adj=1,cex=1.4,font=2,
         labels=c("Across all","Cell type"))
    mtext(paste("DE: |Log2FC| >=",LFC,"& FDR <=",FDR),
          side=3,line=0.2,at=2.5,adj=0,font=2)
    
    par(mar=c(5,0,2,0.5))
    temp <- mapply(function(X,Y,N) {
      plot(NA,NA,xlim=range(X),ylim=range(temp_y),
           xaxt="n",yaxt="n",ylab=NA,xlab=NA)
      temp_X <- X; temp_Y <- Y
      temp_X[temp_X == 0] <- -abs(jitter(temp_X[temp_X == 0],amount=abs(par("usr")[1]) / 2))
      temp_Y[temp_Y == 0] <- -abs(jitter(temp_Y[temp_Y == 0],amount=abs(par("usr")[3]) / 2))
      points(temp_X,temp_Y,pch=19,cex=1.5,
             col=lig_col[names(temp_Y)])
      Z <- Y >= -log10(0.05)
      if (!any(Z)) { Z <- Y > 0 }
      temp_labels <- scClustViz::spreadLabels2(temp_X[Z],temp_Y[Z],label=names(temp_Y)[Z],
                                               str.cex=1.5,str.font=2)
      text(temp_labels,labels=names(temp_Y)[Z],cex=1.2,font=2,
           col=alpha(lig_col[names(temp_Y)[Z]],1))
      axis(1,cex.axis=1.2)
      mtext("# overlapping DE",side=1,line=2.5)
      mtext(N,side=1,line=4,font=2)
    },X=temp_x,Y=temp_y,N=c("Across cell types","Within cell type"))
    rm(list=temp_ds)
    # rm(list=grep("^temp",ls(),value=T))
  }
}
```

```{r AvC_diffexp_boxplot_v3, fig.height=4,fig.width=9}
charLFC <- "4"; charFDR <- "01"
load(paste0("~/Dropbox/GDB_archive/CMapCorr_files/NN_ALLvCT_DEovlp_",
            "LFC",charLFC,"FDR",charFDR,".RData"))
temp_y <- lapply(c(ALL="P_all",CT="P_ct"),function(X) -log10(unlist(get(X))))
temp_x <- list(ALL=sapply(de_all,length),
               CT=sapply(unlist(de_ct,F,T),length))
temp_CTnames <- names(temp_y$CT)
names(temp_y$CT) <- sapply(strsplit(temp_CTnames,".",fixed=T),"[[",1)
names(temp_y$CT)[names(temp_y$CT) %in% unique(names(temp_y$CT)[duplicated(names(temp_y$CT))])] <- 
  temp_CTnames[names(temp_y$CT) %in% unique(names(temp_y$CT)[duplicated(names(temp_y$CT))])]
names(temp_x$CT) <- names(temp_y$CT)

Plist <- Pstat <- list()
for (charLFC in c("2","4")) {
  for (charFDR in c("1","01")) {
    temp_ds <- load(paste0("~/Dropbox/GDB_archive/CMapCorr_files/NN_ALLvCT_DEovlp_",
                        "LFC",charLFC,"FDR",charFDR,".RData"))
    Plist[[length(Plist) + 1]] <- lapply(c(ALL="P_all",CT="P_ct"),function(X) -log10(unlist(get(X))))
    Pstat[[length(Pstat) + 1]] <- wilcox.test(unlist(P_all),unlist(P_ct))$p.value
    names(Plist)[length(Plist)] <- paste("|Log2FC| \u2265",
                                         log2(as.integer(charLFC)),
                                         "& FDR \u2264",
                                         as.numeric(paste0("0.",charFDR)))
    rm(list=temp_ds)
  }
}


par(mfrow=c(1,2),mar=c(3,3,0.5,0.1),mgp=2:0)
plot(NA,NA,xlim=range(unlist(temp_x)),ylim=range(unlist(Plist)),
     yaxt="n",ylab="P # DE overlapping by chance",
     xlab=paste("# overlapping DE at |Log2FC| \u2265",
                log2(as.integer(charLFC)),
                "& FDR \u2264",
                as.numeric(paste0("0.",charFDR))))
axis(2,at=-log10(c(1,0.1,0.05,1e-2,1e-3,1e-4,1e-5)),
     labels=c("100%","10%","5%","1%","0.1%","0.01%","0.001%"))
points(jitter(temp_x$ALL,2),jitter(temp_y$ALL,2),pch=20,col=alpha("red",0.5))
points(jitter(temp_x$CT,2),jitter(temp_y$CT,2),pch=20,col=alpha("dodgerblue",0.5))
text(temp_x$ALL[temp_y$ALL > -log10(0.05)],temp_y$ALL[temp_y$ALL > -log10(0.05)],
     names(temp_y$ALL)[temp_y$ALL > -log10(0.05)],col="red",cex=0.6,pos=3,offset=0.2)
temp_lab1 <- data.frame(x=temp_x$CT[temp_y$CT > -log10(0.05)],
                        y=temp_y$CT[temp_y$CT > -log10(0.05)])
temp_lab2 <- temp_lab1["TGFB1.fibroblast",,drop=F]
temp_lab1 <- temp_lab1[rownames(temp_lab1) != "TGFB1.fibroblast",]
text(temp_lab1,labels=rownames(temp_lab1),
     col="dodgerblue",cex=0.6,offset=0.2,
     pos=c(4,2)[(rownames(temp_lab1) == "BDNF") + 1])
text(temp_lab2,labels=rownames(temp_lab2),
     col="dodgerblue",cex=0.6,offset=0.2,adj=c(-0.01,-0.2))
legend("topright",bty="n",pch=20,col=alpha(c("red","dodgerblue"),0.5),
       legend=c("Across cell types","Within cell type"),cex=0.9,pt.cex=1,x.intersp=0.8)

par(mar=c(3,0.1,0.5,3))
boxplot(unlist(Plist,F,F),pch=20,yaxt="n",xaxt="n",ylab=NA,
        at=seq(1,12)[-seq(3,by=3,length.out=4)],
        col=alpha(c("red","dodgerblue"),0.5),
        border=c("red","dodgerblue"))
axis(4,at=-log10(c(1,0.1,0.05,1e-2,1e-3,1e-4,1e-5)),
     labels=c("100%","10%","5%","1%","0.1%","0.01%","0.001%"))
mtext("P # DE overlapping by chance",side=4,line=2)
mtext(sub("&","\n&",names(Plist)),side=1,line=1,
      at=seq(1.5,by=3,length.out=4),cex=0.9)
mtext(paste("p =",signif(unlist(Pstat),2)),side=1,line=2,
      at=seq(1.5,by=3,length.out=4),cex=0.9)

```


```{r cleanup_AvC}
rm(list=ls())
source("~/Dropbox/GDB/line2user.R")
```


****

# Using all data (including time-series)

NicheNet calculation of time-series data seems to result in all positive logFC values, so that makes some things a little screwy.  In those cases, time-series data was only compared to other time-series data.  

```{r load_data}
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_all_dat.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_all_corr.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_all_DEovlp.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_all_DEmeanlfc.RData")
```

## Correlation

```{r corr_boxplot,fig.height=4,fig.width=9}
par(mar=c(5,3,1,1),mgp=2:0)
temp <- unlist(sapply(nn_ligands,
                      function(X) 
                        list(ALL=corr_all[[X]],
                             CT=corr_ct[[X]],
                             DS=corr_ds[[X]]),
                      simplify=F),recursive=F)
plot(NA,NA,xlim=c(0,length(temp)+1),ylim=range(temp),xaxt="n",xaxs="i",
     xlab=NA,ylab="Spearman correlation of logFC")
boxplot(temp,xaxt="n",yaxt="n",add=T)
mtext(names(temp),side=1,line=0.1,at=seq_along(temp),adj=1,las=3,cex=0.8)
abline(v=seq(3.5,by=3,length.out=length(nn_ligands)-1),col="grey50")

rm(list=grep("^temp",ls(),value=T))
```

```{r corr_boxplot_compTS,fig.height=4,fig.width=9}
par(mfrow=c(1,2),mar=c(3,3,1,1),mgp=2:0)
boxplot(list(ALL=unlist(corr_all),CT=unlist(corr_ct),DS=unlist(corr_ds)),
        ylab="Spearman correlation of logFC")

corr_all_nts <- corr_all_ts <- list()
for (LIG in names(corr_all)) {
  temp_ts <- sapply(strsplit(names(corr_all[[LIG]]),".",fixed=T),
                    function(X) all(DSinfo[[LIG]][X,"time_series"]))
  corr_all_ts[[LIG]] <- corr_all[[LIG]][temp_ts]
  corr_all_nts[[LIG]] <- corr_all[[LIG]][!temp_ts]
}
boxplot(list(ALL_nTS=unlist(corr_all_nts),ALL_TS=unlist(corr_all_ts),
             CT=unlist(corr_ct),DS=unlist(corr_ds)))

rm(list=c("LIG",grep("^temp",ls(),value=T)))
```
Despite the goofiness of the time-series logFC calculations (all are positive?), there's no discernible difference in correlation of logFC within time-series experiments vs non-time-series.  

Note that calculating correlation between time-series and non-TS logFC values is chaos, so those comparisons are skipped in the following figure.  

### Final correlation figure

```{r corr_plot,fig.height=4,fig.width=9}
temp_order <- names(sort(sapply(unique(c(names(corr_all),names(corr_ct),names(corr_ds))),
                                function(X) mean(c(corr_all[[X]],corr_ct[[X]],corr_ds[[X]])))))
lig_col <- qualitative_hcl(length(nn_ligands),palette="dark3",alpha=0.5)
temp_colnames <- c(temp_order,nn_ligands[!nn_ligands %in% temp_order])
names(lig_col) <- temp_colnames[unlist(sapply(seq(1,ceiling(length(nn_ligands)/2)),
                                              function(X) c(X,X + ceiling(length(nn_ligands)/2)),
                                              simplify=F))]
temp_corr <- list(ALL=corr_all[temp_order[temp_order %in% names(corr_all)]],
               CT=corr_ct[temp_order[temp_order %in% names(corr_ct)]],
               DS=corr_ds[temp_order[temp_order %in% names(corr_ds)]])
temp_lig <- sapply(strsplit(names(unlist(temp_corr,recursive=F)),".",fixed=T),"[[",2)


layout(rbind(1:2),widths=c(4,1))

par(mar=c(5,3,1,0.5),mgp=2:0)
plot(NA,NA,xlim=c(1,length(unlist(temp_corr,recursive=F))),ylim=range(temp_corr),
     xaxt="n",xlab=NA,ylab="Spearman correlation of log(fold-change)")
abline(v=cumsum(sapply(temp_corr,length)[1:2]) + 0.5,col="grey70")
abline(h=0,lty=2,col="grey50")
temp_x <- unlist(mapply(rep,x=seq_along(unlist(temp_corr,recursive=F)),
                        times=sapply(unlist(temp_corr,recursive=F),length)))
points(x=jitter(temp_x,amount=0.2),y=unlist(temp_corr),
       pch=19,col=lig_col[temp_lig][temp_x])
mtext(temp_lig,side=1,line=0.1,at=seq_along(temp_lig),las=3,col=alpha(lig_col[temp_lig],1))

segments(x0=c(0,cumsum(sapply(temp_corr,length)[1:2])) + 0.75,
         x1=cumsum(sapply(temp_corr,length)) + 0.25,
         y0=line2user(4,1),y1=line2user(4,1),xpd=NA,lwd=2)
segments(x0=c(c(0,cumsum(sapply(temp_corr,length)[1:2])) + 0.75,
              cumsum(sapply(temp_corr,length)) + 0.25),
         x1=c(c(0,cumsum(sapply(temp_corr,length)[1:2])) + 0.75,
              cumsum(sapply(temp_corr,length)) + 0.25),
         y0=line2user(4,1),y1=line2user(3.8,1),xpd=NA,lwd=2)
mtext(c("Pairwise across cell types / labs","Within same cell type","Same lab"),
      at=0.5 + cumsum(sapply(temp_corr,length)) - sapply(temp_corr,length) / 2,
      side=1,line=4)


par(mar=c(5,0,1,3))
boxplot(list(ALL=unlist(corr_all),CT=unlist(corr_ct),DS=unlist(corr_ds)),
        xaxt="n",xlab=NA,yaxt="n",ylab=NA)
axis(4)
mtext("Spearman correlation of log(fold-change)",side=4,line=2)
mtext(c("Across all","Cell type","Lab"),side=1,line=0.1,at=1:3,las=3)

mtext(paste0("p=",c(signif(wilcox.test(unlist(corr_all),unlist(corr_ct))$p.value,2),
                      signif(wilcox.test(unlist(corr_ct),unlist(corr_ds))$p.value,2))),
      side=1,line=-0.1,at=c(1.5,2.5),las=3,adj=0,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```



## Differential gene expression

### Overlap of DE genes  

Differentially expressed genes were determined by logFC >= 1 and FDR <= 0.1 (as in Browaeys *et al.*, 2020).  This figure shows number of DE genes overlapping across all datasets testing the same ligand vs. datasets testing the same ligand in the same cell type vs. datasets from the same accession number (and same ligand / cell type).  P-values for overlap (y-axis) were determined after determining probability of overlapping DE genes by chance through bootstrapping.

```{r diffexp_boxplot,fig.height=4,fig.width=9}
par(mar=c(1.5,3,1,1),mgp=2:0)
temp_bxp <- list(ALL=-log10(unlist(P_all)),
                 CT=-log10(unlist(P_ct)),
                 DS=-log10(unlist(P_ds)))
boxplot(temp_bxp,ylim=range(temp_bxp),outline=F,
        ylab="-log10(P # DE overlapping by chance)",xaxt="n")
mtext(c("Pairwise across cell types / labs","Within same cell type","Same lab"),
      side=1,line=0.1,at=1:3)
temp <- c("P_all","P_ct","P_ds")
for (X in temp) {
  temp_DE <- -log10(unlist(get(X)))
  names(temp_DE) <- unlist(mapply(rep,x=names(get(X)),times=sapply(get(X),length)))
  temp_DE <- sort(temp_DE)
  temp_x <- seq(which(temp == X) - 0.3,which(temp == X) + 0.3,length.out=length(temp_DE))
  points(temp_x,temp_DE,pch=19,col=lig_col[names(temp_DE)])
  Z <- temp_DE >= -log10(0.05)
  if (!any(Z)) { 
    Z <- temp_DE > 0
  }
  if (any(Z)) { 
    text(temp_x[Z],temp_DE[Z],pos=c(2,4),
         labels=names(temp_DE)[Z],font=2,
         col=alpha(lig_col[names(temp_DE)[Z]],1))
  }
}
mtext(paste("p =",c(signif(wilcox.test(unlist(P_all),unlist(P_ct))$p.value,2),
                    signif(wilcox.test(unlist(P_ct),unlist(P_ds))$p.value,2))),
      side=3,line=-0.9,at=c(1.5,2.5))

rm(list=c("X","Z",grep("^temp",ls(),value=T)))
```

```{r diffexp_boxplot_v2,fig.height=4,fig.width=9}
temp_y <- lapply(c("P_all","P_ct","P_ds"),function(X) {
  temp <- -log10(unlist(get(X)))
  names(temp) <- sapply(strsplit(names(unlist(get(X))),".",fixed=T),"[[",1)
  return(temp)
})
temp_x <- sapply(list(ALL=de_all,CT=de_ct,DS=de_ds),function(DE)
  sapply(unlist(DE,recursive=F,use.names=F),length),simplify=F)

par(mfrow=c(1,4),mar=c(5,4,1,0.5),mgp=2:0)
boxplot(temp_y,ylim=range(temp_y),outline=T,xaxt="n",yaxt="n",ylab=NA)
text(c(1.5,2.5),line2user(-0.2,3),srt=90,adj=1,cex=1.2,
     labels=paste("p =",c(signif(wilcox.test(unlist(P_all),unlist(P_ct))$p.value,2),
                          signif(wilcox.test(unlist(P_ct),unlist(P_ds))$p.value,2))))
axis(2,cex.axis=1.2)
mtext("-log10(P # DE overlapping by chance)",side=2,line=2.5)
text(1:3,line2user(0.5,1),xpd=NA,srt=45,adj=1,cex=1.4,font=2,
     labels=c("Across all","Cell type","Lab"))

par(mar=c(5,0,1,0.5))
temp <- mapply(function(X,Y,N) {
  plot(NA,NA,xlim=range(X),ylim=range(temp_y),
       xaxt="n",yaxt="n",ylab=NA,xlab=NA)
  temp_X <- X; temp_Y <- Y
  temp_X[temp_X == 0] <- -abs(jitter(temp_X[temp_X == 0],amount=abs(par("usr")[1]) / 2))
  temp_Y[temp_Y == 0] <- -abs(jitter(temp_Y[temp_Y == 0],amount=abs(par("usr")[3]) / 2))
  points(temp_X,temp_Y,pch=19,cex=1.5,
         col=lig_col[names(temp_Y)])
  Z <- Y >= -log10(0.05)
  if (!any(Z)) { Z <- Y > 0 }
  temp_labels <- scClustViz::spreadLabels2(temp_X[Z],temp_Y[Z],label=names(temp_Y)[Z],
                                           str.cex=1.5,str.font=2)
  text(temp_labels,labels=names(temp_Y)[Z],cex=1.2,font=2,
       col=alpha(lig_col[names(temp_Y)[Z]],1))
  axis(1,cex.axis=1.2)
  mtext("# overlapping DE",side=1,line=2.5)
  mtext(N,side=1,line=4,font=2)
},X=temp_x,Y=temp_y,N=c("Across cell types / labs","Within same cell type","Same lab"))
```


### Averaged logFC per ligand  

Similar to CMap analyses, took mean LogFC across all samples treated with the same ligand, or same ligand in the same cell line, or from the same accession number (and same ligand / cell line).  

```{r mean_LFC_hist,fig.height=6,fig.width=5}
temp_LFC <- list(ALL=unlist(sapply(mean_lfc,"[[","nts_lfc")),
                 CT=unlist(sapply(unlist(mean_lfc_CT,recursive=F),"[[","lfc")),
                 DS=unlist(sapply(unlist(mean_lfc_DS,recursive=F),"[[","lfc")))
temp_range <- range(unlist(temp_LFC))

temp_FDR <- list(ALL=unlist(sapply(mean_lfc,"[[","nts_fdr")),
                 CT=unlist(sapply(unlist(mean_lfc_CT,recursive=F),"[[","fdr")),
                 DS=unlist(sapply(unlist(mean_lfc_DS,recursive=F),"[[","fdr")))


par(mfrow=c(3,1),mar=c(4,3,1,1),mgp=2:0)
for (X in names(temp_FDR)) {
  temp_main <- switch(X,
                      ALL="Mean LogFC per ligand",
                      CT="By cell type",
                      DS="By dataset")
  
  # hist(temp_LFC[[X]],freq=F,xlim=temp_range,xlab="LogFC",main=temp_main)

  hist(temp_FDR[[X]],freq=F,xlab="FDR",main=temp_main)
  abline(v=mean(temp_FDR[[X]]),lwd=2,col="red")
  if (X == "ALL") {
      legend("topleft",bty="n",lwd=2,col="red",legend="Mean")
  } else {
    legend("topleft",bty="n",lwd=c(2,NA,NA),col=c("red",NA,NA),
           legend=c("Mean","Wilcoxon rank-sum",
                    paste("p =",signif(wilcox.test(temp_FDR[[X]],temp_FDR[[Y]])$p.value,2))))
  }
  # hist(-log10(temp_FDR[[X]]),freq=F,xlim=c(0,temp_max),xlab="-log10(FDR)",main=temp_main)
  # abline(v=mean(-log10(temp_FDR[[X]])),lwd=2,col="red")
  # legend("topright",bty="n",legend="Mean",lwd=2,col="red")
  Y <- X
}

```


### Averaged logFC per ligand compared to CMap  

```{r DEdotplot_fx}
DE_dotplot <- function(LIG,GENES,exclude=c()) {
  temp_DS <- rownames(DSinfo[[LIG]])
  temp_DS <- temp_DS[!temp_DS %in% exclude]
  
  temp_ord_ds <- hclust(dist(t(nn_lig_rep[[LIG]]$lfc[GENES,temp_DS])),"ward.D2")$order
  temp_ord_gene <- hclust(dist(nn_lig_rep[[LIG]]$lfc[GENES,temp_DS]),"ward.D2")$order
  
  CTcol <- qualitative_hcl(length(unique(DSinfo[[LIG]]$cell_type)),palette="dark3")
  names(CTcol) <- unique(DSinfo[[LIG]]$cell_type)
  
  ogLFC <- nn_lig_rep[[LIG]]$lfc[GENES,temp_DS]
  ogLFC <- ogLFC[temp_ord_gene,temp_ord_ds]
  temp_dsname <- DSinfo[[LIG]][colnames(ogLFC),"CtAcc"]
  temp_dsname[DSinfo[[LIG]][colnames(ogLFC),"time_series"]] <-
    paste(temp_dsname[DSinfo[[LIG]][colnames(ogLFC),"time_series"]],"*")
  temp_labelcol <- CTcol[DSinfo[[LIG]][colnames(ogLFC),"cell_type"]]
  names(temp_labelcol) <- colnames(ogLFC) <- temp_dsname
  if (!is.null(mean_lfc[[LIG]])) {
    LFC <- cbind(ogLFC,rep(0,nrow(ogLFC)))
    temp_labelcol <- append(temp_labelcol,"black")
    if (!is.null(mean_lfc[[LIG]]$nts_lfc)) {
      LFC <- cbind(LFC,"Mean (no *)"=mean_lfc[[LIG]][GENES[temp_ord_gene],"nts_lfc"])
      temp_labelcol <- append(temp_labelcol,"black")
    }
    if (!is.null(mean_lfc[[LIG]]$ts_lfc)) {
      LFC <- cbind(LFC,"Mean (only *)"=mean_lfc[[LIG]][GENES[temp_ord_gene],"ts_lfc"])
      temp_labelcol <- append(temp_labelcol,"black")
    }
    if (!is.null(mean_lfc_CT[[LIG]])){
      LFC <- cbind(LFC,sapply(mean_lfc_CT[[LIG]],function(X) X[GENES[temp_ord_gene],"lfc"]))
      temp_labelcol <- append(temp_labelcol,CTcol[names(mean_lfc_CT[[LIG]])])
    }
    if (!is.null(mean_lfc_DS[[LIG]])){
      LFC <- cbind(LFC,sapply(mean_lfc_DS[[LIG]],function(X) X[GENES[temp_ord_gene],"lfc"]))
      temp_labelcol <- append(temp_labelcol,temp_labelcol[names(mean_lfc_DS[[LIG]])])
    }
  } else {
    LFC <- ogLFC
  }
  LFC <- t(matrix(
    cut(c(-1,1,as.vector(LFC / max(LFC))),
        breaks=1000,labels=F)[-1*(1:2)],
    nrow(LFC)))
  
  
  ogFDR <- nn_lig_rep[[LIG]]$qval[GENES,temp_DS]
  ogFDR <- ogFDR[temp_ord_gene,temp_ord_ds]
  colnames(ogFDR) <- temp_dsname
  if (!is.null(mean_lfc[[LIG]])) {
    FDR <- cbind(ogFDR,rep(1,nrow(ogFDR)))
    if (!is.null(mean_lfc[[LIG]]$nts_lfc)) {
      FDR <- cbind(FDR,"Mean (no *)"=mean_lfc[[LIG]][GENES[temp_ord_gene],"nts_fdr"])
    }
    if (!is.null(mean_lfc[[LIG]]$ts_lfc)) {
      FDR <- cbind(FDR,"Mean (only *)"=mean_lfc[[LIG]][GENES[temp_ord_gene],"ts_fdr"])
    }
    if (!is.null(mean_lfc_CT[[LIG]])){
      FDR <- cbind(FDR,sapply(mean_lfc_CT[[LIG]],function(X) X[GENES[temp_ord_gene],"fdr"]))
    }
    if (!is.null(mean_lfc_DS[[LIG]])){
      FDR <- cbind(FDR,sapply(mean_lfc_DS[[LIG]],function(X) X[GENES[temp_ord_gene],"fdr"]))
    }
  } else {
    FDR <- ogFDR
  }
  FDR <- -log10(FDR)
  FDR[FDR > 2] <- 2
  FDR <- t(FDR / max(FDR))
  FDR[FDR == 0] <- NA
  
  if (!any(DSinfo[[LIG]][temp_DS,"time_series"])) {
    rownames(LFC)[rownames(LFC) == "Mean (no *)"] <- "Mean"
    rownames(FDR)[rownames(FDR) == "Mean (no *)"] <- "Mean"
  }
  
  DE <- t(ogLFC >= 1 & ogFDR <= 0.1)
  if (!is.null(mean_lfc[[LIG]])) {
    DE <- rbind(DE,rep(F,ncol(DE)))
    if (!is.null(mean_lfc[[LIG]]$nts_fdr)) {
      DE <- rbind(DE,mean_lfc[[LIG]][GENES[temp_ord_gene],"nts_lfc"] >= 1 & 
                    mean_lfc[[LIG]][GENES[temp_ord_gene],"nts_fdr"] <= 0.1)
    }
    if (!is.null(mean_lfc[[LIG]]$ts_fdr)) {
      DE <- rbind(DE,mean_lfc[[LIG]][GENES[temp_ord_gene],"ts_lfc"] >= 1 & 
                    mean_lfc[[LIG]][GENES[temp_ord_gene],"ts_fdr"] <= 0.1)
    }
    if (!is.null(mean_lfc_CT[[LIG]])){
      DE <- rbind(DE,
                  t(sapply(mean_lfc_CT[[LIG]],function(X) 
                    X[GENES[temp_ord_gene],"lfc"] >= 1 & 
                      X[GENES[temp_ord_gene],"fdr"] <= 0.1)))
    }
    if (!is.null(mean_lfc_DS[[LIG]])){
      DE <- rbind(DE,
                  t(sapply(mean_lfc_DS[[LIG]],function(X) 
                    X[GENES[temp_ord_gene],"lfc"] >= 1 & 
                      X[GENES[temp_ord_gene],"fdr"] <= 0.1)))
    }
  }
  

  par(mar=c(8,7,1,1),mgp=2:0)
  plot(x=NULL,y=NULL,xlim=c(0.5,nrow(LFC) + .5),ylim=c(0.5,ncol(LFC) + .5),
       xaxs="i",yaxs="i",xaxt="n",yaxt="n",xlab=NA,ylab=NA,bty="n",asp=1)
  # abline(v=1:nrow(LFC),col="grey90")
  # abline(h=1:ncol(LFC),col="grey90")
  
  temp_FG <- temp_BG <- diverging_hcl(1000,palette="Blue-Red3")[as.vector(LFC)]
  temp_FG[as.vector(DE)] <- "mediumseagreen"
  symbols(x=rep(1:nrow(FDR),ncol(FDR)),
          y=as.vector(sapply(1:ncol(FDR),function(X) rep(X,nrow(FDR)))),
          circles=as.vector(FDR)/2,inches=F,add=T,xpd=NA,
          fg=temp_FG,bg=temp_BG)

  mtext(colnames(FDR),side=2,at=seq_along(colnames(FDR)),las=2,line=0.1,cex=0.8)
  text(x=seq_along(rownames(FDR)),y=rep(par("usr")[3] - 0.5,nrow(FDR)),
       labels=rownames(FDR),col=temp_labelcol,xpd=NA,adj=1,srt=45,cex=0.8)
  if (any(DSinfo[[LIG]][temp_DS,"time_series"])) {
    text(x=0,y=par("usr")[3] - 0.5,labels="* Time-series dataset",
         xpd=NA,adj=1,srt=45,cex=0.8)
  }
  
  temp_yh <- par("usr")[4] - par("usr")[3]
  segments(x0=rep(line2user(5,2),1000),x1=rep(line2user(4,2),1000),
           y0=seq(0.6 * temp_yh + par("usr")[3],
                  0.9 * temp_yh + par("usr")[3],
                  length.out=1000),
           y1=seq(0.6 * temp_yh + par("usr")[3],
                  0.9 * temp_yh + par("usr")[3],
                  length.out=1000),
           xpd=NA,col=diverging_hcl(1000,palette="Blue-Red3"))
  rect(xleft=line2user(5,2),xright=line2user(4,2),
       ytop=0.9 * temp_yh + par("usr")[3],
       ybottom=(((0.9 - 0.75) * (1 / max(ogLFC))) + 0.75) * temp_yh + par("usr")[3],
       xpd=NA,border="mediumseagreen")
  mtext(c(signif(max(abs(ogLFC)),2),"logFC",-1 * signif(max(abs(ogLFC)),2)),
        side=2,line=c(4,5.1,4),adj=c(-0.1,0.5,1.1),
        at=c(0.9,0.75,0.6) * temp_yh + par("usr")[3])
  
  temp_q <- rev(c(0.5,0.2,0.1,0.05,0.01))
  symbols(x=rep(line2user(4.5,2),length(temp_q)),
          y=seq(0.4 * temp_yh + par("usr")[3],
                by=-1.5,length.out=length(temp_q)),
          circles=-log10(temp_q)/4,inches=F,add=T,xpd=NA,
          fg=ifelse(temp_q <= 0.1,"mediumseagreen","black"))
  mtext(c(paste0("\u2264",temp_q[1] * 100,"%"),
          paste0(temp_q[2:5] * 100,"%")),
        side=2,line=5,cex=0.8,
        at=seq(0.4 * temp_yh + par("usr")[3],
               by=-1.5,length.out=length(temp_q)))
  mtext("FDR",side=2,line=5.8,
        at=0.4 * temp_yh + par("usr")[3] - 3)
  
  mtext(LIG,side=3,line=-0.5,at=line2user(4.5,2),font=2,cex=1.5)
}


DE_dotplot_forCMap <- function(ogFDR,MAIN) {
  FDR <- ogFDR[,colnames(ogFDR) %in% names(mean_lfc)]
  FDR <- FDR[apply(FDR,1,function(X) any(X <= 0.1)),,drop=F]
  # temp_hROW <- hclust(dist(t(FDR)),method="ward.D2") 
  temp_hROW <- list(order=order(colnames(FDR),decreasing=T))
  temp_hGENE <- hclust(dist(FDR),method="ward.D2")
  FDR <- FDR[temp_hGENE$order,temp_hROW$order]
  FDR <- -log10(FDR)
  FDR[FDR > 2] <- 2
  FDR <- FDR / max(FDR)
  FDR[FDR == 0] <- NA
  
  ogZS <- sapply(colnames(FDR),function(LIG)
    rowMeans(lvl4_data@mat[rownames(FDR),lvl4_data@cdesc$pert_iname == LIG]))
  ZS <- matrix(cut(c(-1,1,as.vector(ogZS / max(abs(ogZS)))),
                   breaks=1000,labels=F)[-(1:2)],nrow=nrow(ogZS))
  
  
  par(mar=c(5,7,2,1),mgp=2:0)
  plot(x=NULL,y=NULL,xlim=c(0.5,nrow(FDR) + .5),ylim=c(0.5,ncol(FDR) + .5),
       xaxs="i",yaxs="i",xaxt="n",yaxt="n",xlab=NA,ylab=NA,bty="n",asp=1)
  abline(h=1:ncol(FDR),col="grey90")
  symbols(x=rep(1:nrow(FDR),ncol(FDR)),
          y=as.vector(sapply(1:ncol(FDR),function(X) rep(X,nrow(FDR)))),
          circles=as.vector(FDR)/2,inches=F,add=T,xpd=NA,
          fg=diverging_hcl(1000,palette="Blue-Red3")[as.vector(ZS)],
          bg=diverging_hcl(1000,palette="Blue-Red3")[as.vector(ZS)])
  mtext(colnames(FDR),side=2,at=seq_along(colnames(FDR)),las=2,line=0.1,cex=0.8)
  mtext("Ligands",side=2,line=2.5,font=2)
  mtext(lvl4_data@rdesc[rownames(FDR),"pr_gene_symbol"],
        side=1,las=2,at=1:nrow(FDR),line=0.1,cex=0.8)
  mtext("Genes differentially expressed",side=1,line=3.5,font=2)
  
  temp_yh <- par("usr")[4] - par("usr")[3]
  segments(x0=rep(line2user(5,2),1000),x1=rep(line2user(4,2),1000),
           y0=seq(0.6 * temp_yh + par("usr")[3],
                  0.9 * temp_yh + par("usr")[3],
                  length.out=1000),
           y1=seq(0.6 * temp_yh + par("usr")[3],
                  0.9 * temp_yh + par("usr")[3],
                  length.out=1000),
           xpd=NA,col=diverging_hcl(1000,palette="Blue-Red3"))
  mtext(c(signif(max(abs(ogZS)),2),"Z-score",-1 * signif(max(abs(ogZS)),2)),
        side=2,line=c(4,5.1,4),adj=c(-0.1,0.5,1.1),
        at=c(0.9,0.75,0.6) * temp_yh + par("usr")[3])
  
  temp_q <- rev(c(0.5,0.2,0.1,0.05,0.01))
  symbols(x=rep(line2user(4.5,2),length(temp_q)),
          y=seq(0.4 * temp_yh + par("usr")[3],
                by=-1.5,length.out=length(temp_q)),
          circles=-log10(temp_q)/4,inches=F,add=T,xpd=NA,
          fg="black")
  mtext(c(paste0("\u2264",temp_q[1] * 100,"%"),
          paste0(temp_q[2:5] * 100,"%")),
        side=2,line=5,cex=0.8,
        at=seq(0.4 * temp_yh + par("usr")[3],
               by=-1.5,length.out=length(temp_q)))
  mtext("FDR",side=2,line=5.8,
        at=0.4 * temp_yh + par("usr")[3] - 3)
  
  mtext(MAIN,side=3,line=0.5,font=2,cex=1.2)
}
```


Connectivity Map data from the same set of ligands assayed in multiple experiments from the NicheNet datasets.

```{r CMap_DE_load,fig.height=5,fig.width=9}
require(cmapR)
temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/lvl4_inputs_allgenes.RData")
lvl4_data <- lvl4_data_all
rm(lvl4_data_all,lig16)

load("~/Dropbox/GDB_archive/CMapCorr_files/lig295_DE_allgenes_lig_FDR.RData")
FDRinf <- FDR_lig
load("~/Dropbox/GDB_archive/CMapCorr_files/lig295_DE_lig_FDR.RData")
FDRass <- FDR_lig
rm(FDR_lig)

par(mfrow=c(1,2))
DE_dotplot_forCMap(FDRass,"CMap (assayed genes only)")
DE_dotplot_forCMap(FDRinf,"CMap (including inferred)")

```


```{r DE_dot}
for (LIG in nn_ligands) {
  # print(paste(LIG,which(nn_ligands == LIG),"/",length(nn_ligands)))
  
  temp_genes <- Reduce(union,nn_DE[[LIG]])
  if (LIG %in% colnames(FDRinf)) {
    temp_cmapgenes <- unique(
      lvl4_data@rdesc[c(rownames(FDRass)[FDRass[,LIG] <= 0.1],
                        rownames(FDRinf)[FDRinf[,LIG] <= 0.1]),"pr_gene_symbol"]
    )
    temp_cmapgenes <- temp_cmapgenes[temp_cmapgenes %in% temp_genes]
  } else {
    temp_cmapgenes <- character()
  }
  # print(paste("cmapgenes =",length(temp_cmapgenes)))
  
  if (!is.null(mean_lfc[[LIG]])) {
    temp_meanhits <- rownames(mean_lfc[[LIG]])[
      apply(mean_lfc[[LIG]][,colnames(mean_lfc[[LIG]]) %in% c("nts_fdr","ts_fdr"),drop=F],1,
            function(X) any(X <= 0.1))
    ]
    # print(paste("mean_lfc =",length(temp_meanhits)))
    if (length(temp_meanhits) < 1) {
      temp_meanhits <- rownames(mean_lfc[[LIG]])[c(which.min(mean_lfc[[LIG]]$nts_fdr),
                                                   which.min(mean_lfc[[LIG]]$ts_fdr))]
    }
  } else {
    temp_meanhits <- character()
  }
  
  temp_DEoverlap <- rowSums(sapply(nn_DE[[LIG]],function(X) temp_genes %in% X))
  # print(table(temp_DEoverlap))
  if (sum(temp_DEoverlap == max(temp_DEoverlap)) > 50) {
    temp_commonDE <- names(sort(rowSums(
      nn_lig_rep[[LIG]]$qval[temp_genes[temp_DEoverlap == max(temp_DEoverlap)],]
    ))[1:50])
  } else if (length(c(temp_meanhits,temp_cmapgenes)) +
             sum(temp_DEoverlap == max(temp_DEoverlap)) < 15) {
    temp_commonDE <- c(temp_genes[temp_DEoverlap == max(temp_DEoverlap)],
                       names(sort(rowSums(nn_lig_rep[[LIG]]$qval[temp_genes,]))[1:10]))
  } else {
    temp_commonDE <- temp_genes[temp_DEoverlap == max(temp_DEoverlap)]
  }

  GENES <- unique(c(temp_commonDE,temp_meanhits,temp_cmapgenes))
  GENES <- GENES[!is.na(GENES)]
  # print(length(GENES))
  # print("")
  
  temp_Ylen <- length(GENES)
  temp_Xlen <- nrow(DSinfo[[LIG]]) + 
    as.integer(!is.null(mean_lfc[[LIG]])) + 
    length(mean_lfc[[LIG]][1,]) / 3 +
    length(mean_lfc_CT[[LIG]]) +
    length(mean_lfc_DS[[LIG]])
  
  Sys.sleep(0.1)
  png(paste0("~/Dropbox/GDB/CMapCorr/docs/output_figs/NN_",LIG,".png"),
      width=temp_Xlen * 0.2 + 1.6,height=temp_Ylen * 0.2 + 1.8,
      units="in",res=150)
  DE_dotplot(LIG,GENES)
  dev.off()
  
  rm(list=c("GENES",grep("^temp",ls(),value=T)))
}
```


```{r DE_dot_noTS}
for (LIG in nn_ligands) {
  if (sum(!DSinfo[[LIG]]$time_series) < 2) { next }
  # print(paste(LIG,which(nn_ligands == LIG),"/",length(nn_ligands)))
  
  temp_genes <- Reduce(union,nn_DE[[LIG]])
  if (LIG %in% colnames(FDRinf)) {
    temp_cmapgenes <- unique(
      lvl4_data@rdesc[c(rownames(FDRass)[FDRass[,LIG] <= 0.1],
                        rownames(FDRinf)[FDRinf[,LIG] <= 0.1]),"pr_gene_symbol"]
    )
    temp_cmapgenes <- temp_cmapgenes[temp_cmapgenes %in% temp_genes]
  } else {
    temp_cmapgenes <- character()
  }
  # print(paste("cmapgenes =",length(temp_cmapgenes)))
  
  if (!is.null(mean_lfc[[LIG]]$nts_fdr)) {
    temp_meanhits <- rownames(mean_lfc[[LIG]])[mean_lfc[[LIG]]$nts_fdr <= 0.1]
    # print(paste("mean_lfc =",length(temp_meanhits)))
    if (length(temp_meanhits) < 1) {
      temp_meanhits <- rownames(mean_lfc[[LIG]])[which.min(mean_lfc[[LIG]]$nts_fdr)]
    }
  } else {
    temp_meanhits <- character()
  }
  mean_lfc[[LIG]] <- mean_lfc[[LIG]][,!grepl("^ts",colnames(mean_lfc[[LIG]]))]
  
  temp_DEoverlap <- rowSums(sapply(nn_DE[[LIG]],function(X) temp_genes %in% X))
  # print(table(temp_DEoverlap))
  if (sum(temp_DEoverlap == max(temp_DEoverlap)) > 50) {
    temp_commonDE <- names(sort(rowSums(
      nn_lig_rep[[LIG]]$qval[temp_genes[temp_DEoverlap == max(temp_DEoverlap)],]
    ))[1:50])
  } else if (length(c(temp_meanhits,temp_cmapgenes)) +
             sum(temp_DEoverlap == max(temp_DEoverlap)) < 15) {
    temp_commonDE <- c(temp_genes[temp_DEoverlap == max(temp_DEoverlap)],
                       names(sort(rowSums(nn_lig_rep[[LIG]]$qval[temp_genes,]))[1:10]))
  } else {
    temp_commonDE <- temp_genes[temp_DEoverlap == max(temp_DEoverlap)]
  }

  GENES <- unique(c(temp_commonDE,temp_meanhits,temp_cmapgenes))
  GENES <- GENES[!is.na(GENES)]
  # print(length(GENES))
  # print("")
  
  temp_Ylen <- length(GENES)
  temp_Xlen <- nrow(DSinfo[[LIG]]) + 
    as.integer(!is.null(mean_lfc[[LIG]])) + 
    length(mean_lfc[[LIG]][1,]) / 3 +
    length(mean_lfc_CT[[LIG]]) +
    length(mean_lfc_DS[[LIG]])
  
  Sys.sleep(0.1)
  png(paste0("~/Dropbox/GDB/CMapCorr/docs/output_figs/NN_",LIG,"_noTS.png"),
      width=temp_Xlen * 0.2 + 1.6,height=temp_Ylen * 0.2 + 1.8,
      units="in",res=150)
  DE_dotplot(LIG,GENES,exclude=rownames(DSinfo[[LIG]])[DSinfo[[LIG]]$time_series])
  dev.off()
  
  rm(list=c("GENES",grep("^temp",ls(),value=T)))
}
```


Modified heatmaps showing logFC and FDR for selected genes from each sample from the NicheNet gold standard, as well as averages across all samples, all cell types, and all datasets where available.  The figure below is an example for TNF.  

```{r dotplot_links,echo=F,results="asis"}
temp_files <- list.files("docs/output_figs/",pattern="^NN")
temp_lig <- matrix(sort(nn_ligands),ncol=4,byrow=T)
for (L in 1:nrow(temp_lig)) {
  cat("<p>")
  for (LIG in temp_lig[L,]) {
    if (!any(grepl(LIG,temp_files))) { next }
    cat(" | ")
    temp_ligfiles <- grep(paste0("_",LIG,"[._]"),temp_files,value=T)
    cat(paste0('<a href="output_figs/',temp_ligfiles[!grepl("noTS",temp_ligfiles)],'">',LIG,'</a>'))
    if (any(grepl("noTS",temp_ligfiles))) {
      cat(paste0('<a href="output_figs/',
                 temp_ligfiles[grepl("noTS",temp_ligfiles)],
                 '">(no time-series)</a>'))
    }
    cat(" | ")
  }
  cat("</p>")
  cat("\n")
}

cat('<p><img src="output_figs/NN_TNF_noTS.png"/></p>')
```


### Comparing NicheNet & CMap expression changes per ligand

```{r CMap_NN_compare}
for (X in names(DSinfo)) {
  DSinfo[[X]]$CtAcc <- paste(DSinfo[[X]]$cell_type,DSinfo[[X]]$accession)
}

for (LIG in sort(nn_ligands)) {
  if (all(DSinfo[[LIG]]$time_series)) { next }
  if (!LIG %in% lvl4_data@cdesc$pert_iname) { next }
  
  # CMap genes ----
  CM_ogZS <- sapply(unique(lvl4_data@cdesc$cell_id[lvl4_data@cdesc$pert_iname == LIG]),function(CT) 
    rowMeans(lvl4_data@mat[,lvl4_data@cdesc$pert_iname == LIG & lvl4_data@cdesc$cell_id == CT]))
  CM_ogFDR <- apply(CM_ogZS,2,function(Z) p.adjust(pnorm(-abs(Z))))
  colnames(CM_ogFDR) <- colnames(CM_ogZS) <- sapply(colnames(CM_ogZS),function(X) names(ct14)[ct14 == X])
  
  CM_GENES <- rownames(CM_ogFDR)[apply(CM_ogFDR,1,function(X) sum(X <= 0.5) >= 2)]
  # CM_GENES <- rownames(CM_ogFDR)[apply(CM_ogFDR,1,function(X) any(X <= 0.1))]
  
  
  # NN genes ----
  NN_ogZS <- nn_lig_rep[[LIG]]$lfc
  # NN_ogZS <- NN_ogZS[,!DSinfo[[LIG]][colnames(NN_ogZS),"time_series"],drop=F]
  # CAN'T DROP TIMESERIES
  NN_ogFDR <- nn_lig_rep[[LIG]]$qval[,colnames(NN_ogZS),drop=F]
  temp_ts <- DSinfo[[LIG]][colnames(NN_ogZS),"time_series"]
  colnames(NN_ogFDR) <- colnames(NN_ogZS) <- paste(DSinfo[[LIG]][colnames(NN_ogZS),"CtAcc"],"*")
  colnames(NN_ogFDR)[temp_ts] <- colnames(NN_ogZS)[temp_ts] <- paste0(colnames(NN_ogZS)[temp_ts],"*")
  # if (LIG == "VEGFA") {
  #   NN_GENES <- character()
  # } else {  
  #   temp_samples <- ncol(NN_ogFDR) + 1
  #   NN_GENES <- character()
  #   while (length(NN_GENES) < 3) {
  #     temp_samples <- temp_samples - 1
  #     if (temp_samples <= 1) { break }
  #     NN_GENES <- rownames(NN_ogZS)[apply(sapply(colnames(NN_ogZS),function(X)
  #       abs(NN_ogZS[,X]) >= 1 & NN_ogFDR[,X] <= 0.01),
  #       1,function(Y) sum(Y) >= temp_samples)]  
  #     print(length(NN_GENES))
  #     # print(temp_samples)
  #   }
  # }
  # if (any(any(!NN_GENES %in% lvl4_data@rdesc$pr_gene_symbol))) {
  #   # print(NN_GENES[!NN_GENES %in% lvl4_data@rdesc$pr_gene_symbol])
  # }
  NN_GENES <- lvl4_data@rdesc[CM_GENES,"pr_gene_symbol"]
  NN_GENES <- NN_GENES[NN_GENES %in% rownames(NN_ogFDR)]
  
  # Merge genes & fix gene names ----
  temp_CMgenes <- GENES <- unique(c(NN_GENES,lvl4_data@rdesc[CM_GENES,"pr_gene_symbol"]))
  # Fixing SELENOP (formerly SEPP1)
  temp_CMgenes[temp_CMgenes == "SELENOP"] <- "SEPP1"
  # Fixing missing genes
  temp_CMgenes <- temp_CMgenes[temp_CMgenes %in% lvl4_data@rdesc$pr_gene_symbol]
  GENES <- GENES[GENES %in% lvl4_data@rdesc$pr_gene_symbol]
  temp_CMgenes <- temp_CMgenes[temp_CMgenes %in% rownames(NN_ogZS)]
  GENES <- GENES[GENES %in% rownames(NN_ogZS)]
  
  temp_CMgenes <- sapply(temp_CMgenes,function(X) 
    rownames(lvl4_data@rdesc)[lvl4_data@rdesc$pr_gene_symbol == X])
  names(temp_CMgenes)[names(temp_CMgenes) == "SEPP1"] <- "SELENOP"
  
  
  # Filter CMap ----
  CM_ogZS <- CM_ogZS[temp_CMgenes,,drop=F]
  CM_ogFDR <- CM_ogFDR[temp_CMgenes,,drop=F]
  rownames(CM_ogZS) <- rownames(CM_ogFDR) <- GENES
  
  
  # Filter NN ----
  NN_ogZS <- NN_ogZS[GENES,,drop=F]
  NN_ogFDR <- NN_ogFDR[GENES,,drop=F]
  
  
  # Merge data ----
  ogZS <- cbind(CM_ogZS,NN_ogZS)
  ogFDR <- cbind(CM_ogFDR,NN_ogFDR)
  
  # HClust ----
  if (nrow(ogZS) > 1) {
    temp_mat <- -log10(ogFDR)
    temp_mat[temp_mat == 0] <- 1e-10
    temp_mat <- temp_mat * ogZS
    temp_corS <- cor(temp_mat,method="pearson")
    temp_corG <- cor(t(temp_mat),method="pearson")
    hS <- hclust(as.dist((1 - abs(temp_corS)) * sign(temp_corS)),method="ward.D2")
    hG <- hclust(as.dist((1 - abs(temp_corG)) * sign(temp_corG)),method="ward.D2")
    ogZS <- ogZS[hG$order,hS$order]
    ogFDR <- ogFDR[hG$order,hS$order]
  }
  
  # Scale CMap ----
  CM_cut <- 5
  # boxplot(CM_ogZS,ylim=c(-CM_cut,CM_cut))
  CM_ZS <- CM_ogZS
  CM_ZS[CM_ZS > CM_cut] <- CM_cut; CM_ZS[CM_ZS < -CM_cut] <- -CM_cut
  CM_ZS <- matrix(cut(c(-1,1,as.vector(CM_ogZS / max(abs(CM_ogZS)))),
                      breaks=1000,labels=F)[-(1:2)],nrow=nrow(CM_ogZS))
  rownames(CM_ZS) <- rownames(CM_ogZS); colnames(CM_ZS) <- colnames(CM_ogZS)
  
  CM_FDR <- -log10(CM_ogFDR)
  CM_FDR[CM_FDR > 3] <- 3
  CM_FDR <- CM_FDR / max(CM_FDR)
  CM_FDR[CM_FDR <= 0] <- NA
  
  
  # Scale NN ----
  NN_cut <- CM_cut # if you want these to be different, change the legend.
  # boxplot(NN_ogZS,ylim=c(-NN_cut,NN_cut))
  NN_ZS <- NN_ogZS
  NN_ZS[NN_ZS > NN_cut] <- NN_cut; NN_ZS[NN_ZS < -NN_cut] <- -NN_cut
  NN_ZS <- matrix(cut(c(-1,1,as.vector(NN_ogZS / max(abs(NN_ogZS)))),
                      breaks=1000,labels=F)[-(1:2)],nrow=nrow(NN_ogZS))
  rownames(NN_ZS) <- rownames(NN_ogZS); colnames(NN_ZS) <- colnames(NN_ogZS)
  
  NN_FDR <- -log10(NN_ogFDR)
  NN_FDR[NN_FDR > 3] <- 3
  NN_FDR <- NN_FDR / max(NN_FDR)
  NN_FDR[NN_FDR <= 0] <- NA
  
  
  # Merge plotting data ----
  ZS <- t(cbind(CM_ZS,NN_ZS)[rownames(ogZS),colnames(ogZS)])
  FDR <- t(cbind(CM_FDR,NN_FDR)[rownames(ogFDR),colnames(ogFDR)])
  

  # PLOT ----
  png(paste0("docs/output_figs/CMapNN_",LIG,".png"),
      width=nrow(FDR) * 0.2 + 2,height=ncol(FDR) * 0.2 + 1.6,
      units="in",res=120)
  
  par(mar=c(7,9,1,1),mgp=2:0)
  plot(x=NULL,y=NULL,xlim=c(0.5,nrow(FDR) + .5),ylim=c(0.5,ncol(FDR) + .5),
       xaxs="i",yaxs="i",xaxt="n",yaxt="n",xlab=NA,ylab=NA,bty="n",asp=1)
  # abline(h=1:ncol(FDR),col="grey90")
  abline(v=1:nrow(FDR),col="grey90")
  abline(h=1:ncol(FDR),col="grey90")
  symbols(x=rep(1:nrow(FDR),ncol(FDR)),
          y=as.vector(sapply(1:ncol(FDR),function(X) rep(X,nrow(FDR)))),
          circles=as.vector(FDR)/2,inches=F,add=T,xpd=NA,
          fg=diverging_hcl(1000,palette="Blue-Red3")[as.vector(ZS)],
          bg=diverging_hcl(1000,palette="Blue-Red3")[as.vector(ZS)])
  mtext("Genes",side=2,line=0.3,at=line2user(0.1,3),las=2,font=2)
  text(rep(line2user(0.3,1),ncol(FDR)),1:ncol(FDR),
       ifelse(as.logical(as.integer(lvl4_data@rdesc[temp_CMgenes[colnames(FDR)],"pr_is_lm"])),
              yes=colnames(FDR),paste0(colnames(FDR),"^")),
       xpd=NA,adj=1,srt=45,cex=0.8)
  text(line2user(0,4),line2user(1.3,2),"Cell lines",xpd=NA,adj=1,srt=45,font=2)
  text(1:nrow(FDR),rep(line2user(0.3,2),nrow(FDR)),
       labels=rownames(FDR),xpd=NA,adj=1,srt=45,cex=0.8)
  
  mtext(LIG,side=3,line=0,at=line2user(5.4,2),font=2,cex=1.2)
  
  temp_yh <- par("usr")[4] - par("usr")[3]
  mtext("Z-score /",side=2,line=6.9,at=0.75 * temp_yh + par("usr")[3])
  mtext("LogFC *",side=2,line=6.1,at=0.75 * temp_yh + par("usr")[3])
  segments(x0=rep(line2user(6,2),1000),x1=rep(line2user(4.8,2),1000),
           y0=seq(0.5 * temp_yh + par("usr")[3],
                  temp_yh + par("usr")[3],
                  length.out=1000),
           y1=seq(0.5 * temp_yh + par("usr")[3],
                  temp_yh + par("usr")[3],
                  length.out=1000),
           xpd=NA,col=diverging_hcl(1000,palette="Blue-Red3"))
  text(rep(line2user(5.4,2),2),c(1,0.5) * temp_yh + par("usr")[3],
       labels=c(paste0("\u2265",CM_cut),paste0("\u2264",-CM_cut)),
       col="grey80",pos=c(1,3),offset=0.1,xpd=NA)
  
  mtext("FDR",side=2,line=6,las=2,
        at=1 + 0.25 * temp_yh + par("usr")[3])
  temp_q <- rev(c(0.2,0.1,0.05,0.01,0.001))
  symbols(x=rep(line2user(5.4,2),length(temp_q)),
          y=seq(0.25 * temp_yh + par("usr")[3],
                by=-1,length.out=length(temp_q)),
          circles=-log10(temp_q) / (max(-log10(temp_q)) *2),
          inches=F,add=T,xpd=NA,fg="black")
  mtext(c(paste0("\u2264",temp_q[1] * 100,"%"),
          paste0(temp_q[2:5] * 100,"%")),
        side=2,line=6,cex=0.8,las=2,
        at=seq(0.25 * temp_yh + par("usr")[3],
               by=-1,length.out=length(temp_q)))
  
  mtext(paste("* microarray data",
              "** timeseries data",
              "^ inferred in CMap",
              sep="; "),
        side=3,line=0.1,cex=0.6,at=line2user(0,4),adj=1)
  
  dev.off()
}
```

Plots per ligand treatment of mean Z-score for each CMap cell line and logFC from NicheNet datasets.  

```{r CMAP_NN_links,echo=F,results="asis"}
temp_files <- list.files("docs/output_figs/",pattern="CMap")
temp_lig <- matrix(sort(nn_ligands),ncol=4,byrow=T)
for (L in 1:nrow(temp_lig)) {
  cat("<p>")
  for (LIG in temp_lig[L,]) {
    if (!any(grepl(LIG,temp_files))) { next }
    cat(" | ")
    temp_ligfiles <- grep(paste0("_",LIG,"[._]"),temp_files,value=T)
    cat(paste0('<a href="output_figs/',temp_ligfiles[!grepl("noTS",temp_ligfiles)],'">',LIG,'</a>'))
    if (any(grepl("noTS",temp_ligfiles))) {
      cat(paste0('<a href="output_figs/',
                 temp_ligfiles[grepl("noTS",temp_ligfiles)],
                 '">(no time-series)</a>'))
    }
    cat(" | ")
  }
  cat("</p>")
  cat("\n")
}

cat('<p><img src="output_figs/CMapNN_IFNG.png"/></p>')
```

```{r load_lvl5, eval=F}
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs_allgenes.RData")
rm(lig16)

Y <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == "TNF"]
X <- rownames(lvl5_data@mat)[unique(apply(lvl5_data@mat[,Y],2,which.max))]



image(lvl5_data@mat[X,Y])

```



****

# After removing time-series data

```{r noTS_load_data}
rm(list=ls())

source("~/Dropbox/GDB/line2user.R")

load("~/Dropbox/GDB_archive/CMapCorr_files/NN_noTS_dat.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_noTS_corr.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_noTS_DEovlp.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/NN_noTS_DEmeanlfc.RData")
```

## Correlation


```{r noTS_corr_plot,fig.height=4,fig.width=9}
temp_order <- names(sort(sapply(unique(c(names(corr_all),names(corr_ct),names(corr_ds))),
                                function(X) mean(c(corr_all[[X]],corr_ct[[X]],corr_ds[[X]])))))
lig_col <- qualitative_hcl(length(temp_order),palette="dark3",alpha=0.5)
names(lig_col) <- temp_order[unlist(sapply(1:5,function(X) c(X,X+5),simplify=F))]
temp_corr <- list(ALL=corr_all[temp_order[temp_order %in% names(corr_all)]],
               CT=corr_ct[temp_order[temp_order %in% names(corr_ct)]],
               DS=corr_ds[temp_order[temp_order %in% names(corr_ds)]])
temp_lig <- sapply(strsplit(names(unlist(temp_corr,recursive=F)),".",fixed=T),"[[",2)


layout(rbind(1:2),widths=c(4,1))

par(mar=c(5,3,1,0.5),mgp=2:0)
plot(NA,NA,xlim=c(1,length(unlist(temp_corr,recursive=F))),ylim=range(temp_corr),
     xaxt="n",xlab=NA,ylab="Spearman correlation of log(fold-change)")
abline(v=cumsum(sapply(temp_corr,length)[1:2]) + 0.5,col="grey70")
abline(h=0,lty=2,col="grey50")
temp_x <- unlist(mapply(rep,x=seq_along(unlist(temp_corr,recursive=F)),
                        times=sapply(unlist(temp_corr,recursive=F),length)))
points(x=jitter(temp_x,amount=0.2),y=unlist(temp_corr),
       pch=19,col=lig_col[temp_lig][temp_x])
mtext(temp_lig,side=1,line=0.1,at=seq_along(temp_lig),las=3,col=alpha(lig_col[temp_lig],1))

segments(x0=c(0,cumsum(sapply(temp_corr,length)[1:2])) + 0.75,
         x1=cumsum(sapply(temp_corr,length)) + 0.25,
         y0=line2user(4,1),y1=line2user(4,1),xpd=NA,lwd=2)
segments(x0=c(c(0,cumsum(sapply(temp_corr,length)[1:2])) + 0.75,
              cumsum(sapply(temp_corr,length)) + 0.25),
         x1=c(c(0,cumsum(sapply(temp_corr,length)[1:2])) + 0.75,
              cumsum(sapply(temp_corr,length)) + 0.25),
         y0=line2user(4,1),y1=line2user(3.8,1),xpd=NA,lwd=2)
mtext(c("Pairwise across cell types / labs","Within same cell type","Same lab"),
      at=0.5 + cumsum(sapply(temp_corr,length)) - sapply(temp_corr,length) / 2,
      side=1,line=4)


par(mar=c(5,0,1,3))
boxplot(list(ALL=unlist(corr_all),CT=unlist(corr_ct),DS=unlist(corr_ds)),
        xaxt="n",xlab=NA,yaxt="n",ylab=NA)
axis(4)
mtext("Spearman correlation of log(fold-change)",side=4,line=2)
mtext(c("Across all","Cell type","Lab"),side=1,line=0.1,at=1:3,las=3)

mtext(paste0("p=",c(signif(wilcox.test(unlist(corr_all),unlist(corr_ct))$p.value,2),
                      signif(wilcox.test(unlist(corr_ct),unlist(corr_ds))$p.value,2))),
      side=1,line=-0.1,at=c(1.5,2.5),las=3,adj=0,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```



## Differential gene expression

```{r noTS_diffexp_boxplot,fig.height=4,fig.width=9}
par(mar=c(1.5,3,1,1),mgp=2:0)
temp_bxp <- list(ALL=-log10(unlist(P_all)),
                 CT=-log10(unlist(P_ct)),
                 DS=-log10(unlist(P_ds)))
boxplot(temp_bxp,ylim=range(temp_bxp),outline=F,
        ylab="-log10(P # DE overlapping by chance)",xaxt="n")
mtext(c("Pairwise across cell types / labs","Within same cell type","Same lab"),
      side=1,line=0.1,at=1:3)
temp <- c("P_all","P_ct","P_ds")
for (X in temp) {
  temp_DE <- -log10(unlist(get(X)))
  names(temp_DE) <- unlist(mapply(rep,x=names(get(X)),times=sapply(get(X),length)))
  temp_DE <- sort(temp_DE)
  temp_x <- seq(which(temp == X) - 0.3,which(temp == X) + 0.3,length.out=length(temp_DE))
  points(temp_x,temp_DE,pch=19,col=lig_col[names(temp_DE)])
  Z <- temp_DE >= -log10(0.05)
  if (!any(Z)) { 
    Z <- temp_DE > 0
  }
  if (any(Z)) { 
    text(temp_x[Z],temp_DE[Z],pos=c(2,4),
         labels=names(temp_DE)[Z],font=2,
         col=alpha(lig_col[names(temp_DE)[Z]],1))
  }
}
mtext(paste("p =",c(signif(wilcox.test(unlist(P_all),unlist(P_ct))$p.value,2),
                    signif(wilcox.test(unlist(P_ct),unlist(P_ds))$p.value,2))),
      side=3,line=-0.9,at=c(1.5,2.5))

rm(list=c("X","Z",grep("^temp",ls(),value=T)))
```

```{r noTS_diffexp_boxplot_v2,fig.height=4,fig.width=9}
temp_y <- lapply(c("P_all","P_ct","P_ds"),function(X) {
  temp <- -log10(unlist(get(X)))
  names(temp) <- sapply(strsplit(names(unlist(get(X))),".",fixed=T),"[[",1)
  return(temp)
})
temp_x <- sapply(list(ALL=de_all,CT=de_ct,DS=de_ds),function(DE)
  sapply(unlist(DE,recursive=F,use.names=F),length),simplify=F)

par(mfrow=c(1,4),mar=c(5,4,1,0.5),mgp=2:0)
boxplot(temp_y,ylim=range(temp_y),outline=T,xaxt="n",yaxt="n",ylab=NA)
text(c(1.5,2.5),line2user(-0.2,3),srt=90,adj=1,cex=1.2,
     labels=paste("p =",c(signif(wilcox.test(unlist(P_all),unlist(P_ct))$p.value,2),
                          signif(wilcox.test(unlist(P_ct),unlist(P_ds))$p.value,2))))
axis(2,cex.axis=1.2)
mtext("-log10(P # DE overlapping by chance)",side=2,line=2.5)
text(1:3,line2user(0.5,1),xpd=NA,srt=45,adj=1,cex=1.4,font=2,
     labels=c("Across all","Cell type","Lab"))

par(mar=c(5,0,1,0.5))
temp <- mapply(function(X,Y,N) {
  plot(NA,NA,xlim=range(X),ylim=range(temp_y),
       xaxt="n",yaxt="n",ylab=NA,xlab=NA)
  temp_X <- X; temp_Y <- Y
  temp_X[temp_X == 0] <- -abs(jitter(temp_X[temp_X == 0],amount=abs(par("usr")[1]) / 2))
  temp_Y[temp_Y == 0] <- -abs(jitter(temp_Y[temp_Y == 0],amount=abs(par("usr")[3]) / 2))
  points(temp_X,temp_Y,pch=19,cex=1.5,
         col=lig_col[names(temp_Y)])
  Z <- Y >= -log10(0.05)
  if (!any(Z)) { Z <- Y > 0 }
  temp_labels <- scClustViz::spreadLabels2(temp_X[Z],temp_Y[Z],label=names(temp_Y)[Z],
                                           str.cex=1.5,str.font=2)
  text(temp_labels,labels=names(temp_Y)[Z],cex=1.2,font=2,
       col=alpha(lig_col[names(temp_Y)[Z]],1))
  axis(1,cex.axis=1.2)
  mtext("# overlapping DE",side=1,line=2.5)
  mtext(N,side=1,line=4,font=2)
},X=temp_x,Y=temp_y,N=c("Pairwise across cell types / labs","Within same cell type","Same lab"))
```

