---
title: "Biological factors affecting transcriptional response to ligand stimulus"
---
  
```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(cmapR)
library(ranger)
library(scales)
library(colorspace)
library(kableExtra)
source("~/Dropbox/GDB/line2user.R")
knitr::opts_chunk$set(
  echo=F,
  message=F,
  warning=F
)
```

Associating receptor mRNA / protein abundance with transcriptional response. The quantile-normalized transcriptomes per cell type (level 3 data) will be used to determine whether the cognate receptor for each ligand is present.  

# Receptor mapping from CCInx

Ligand-receptor interactions are from the Bader lab database included in [CCInx](https://baderlab.github.io/CCInx/).

## Receptor expression per cell line  

```{r load_data_lvl3_ctl_ccinx,include=TRUE}
if (exists("lvl3_ctl")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData") 
} else {
  source("lvl3_inputs_allgenesctrl.R")
}

temp <- load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_human.RData",package="CCInx"))

lig16R <- sapply(lig16,function(X) {
  temp <- c(inxDB$nodeB[inxDB$nodeA == X],inxDB$nodeA[inxDB$nodeB == X])
  return(temp[grepl("Receptor",geneInfo[temp,"protein_type"])])
},simplify=F)
lig16R[["GAS6"]] <- c("AXL","TYRO3","MERTK")
for (LIG in lig16) {
  temp_in <- lig16R[[LIG]] %in% lvl3_ctl@rdesc$pr_gene_symbol
  lig16R[[LIG]] <- lig16R[[LIG]][temp_in]
  names(lig16R[[LIG]]) <- sapply(lig16R[[LIG]],function(X) 
    lvl3_ctl@rdesc$id[lvl3_ctl@rdesc$pr_gene_symbol == X])
  cat(paste0(round(sum(temp_in) / length(temp_in) * 100),"% (",
               round(sum(lvl3_ctl@rdesc[names(lig16R[[LIG]]),"pr_is_lm"] == "1") / length(temp_in) * 100),
               "%) of ",LIG," receptors present in CMap data, either inferred or (assayed)."),"\n")
}

rm(list=c("LIG","temp",temp))
```

Below is an overview of all potential cognate receptors per ligand present in the data.  The left figure shows which ligands (dark purple) each receptor is known to interact with.  On the right is a heatmap showing mean normalized gene expression per cell type for each receptor gene, with darker indicating higher expression.  Gene names (rows) are ordered by heirarchical clustering of ligand-receptor interaction data, such that receptors common to the same ligand should be together.  **Landmark** gene expression was emperically determined rather than inferred (names indicated in bold).  Ligand names (columns, left) are ordered by heirarchical clustering of ligand-receptor interaction data, such that ligands sharing the same receptors should be proximal.  Cell type names (columns, right) are ordered by heirarchical clustering of mean receptor gene expression.

```{r rec_expr_ccinx,fig.height=25,fig.width=9}
temp_genes <- unique(unlist(lapply(lig16R,names)))
names(temp_genes) <- lvl3_ctl@rdesc[temp_genes,"pr_gene_symbol"]
lig16Rmat <- sapply(lig16R,function(X) names(temp_genes) %in% X)
rownames(lig16Rmat) <- names(temp_genes)
temp_roworder <- hclust(dist(lig16Rmat))$order
lig16Rmat <- lig16Rmat[temp_roworder,]
lig16Rmat <- lig16Rmat[,hclust(dist(t(lig16Rmat)))$order]

meanR <- sapply(ct14,function(CT) rowMeans(lvl3_ctl@mat[temp_genes,lvl3_ctl@cdesc$cell_id == CT]))
rownames(meanR) <- names(temp_genes)
meanR <- meanR[temp_roworder,]
meanR <- meanR[,hclust(dist(t(meanR)))$order]

layout(cbind(rep(1,25),rep(2,25),c(rep(3,5),rep(0,20))),widths=c(3,3,1))
par(mar=c(6,0.5,6,5.5),mgp=2:0,las=2)
image(z=t(lig16Rmat[rev(rownames(lig16Rmat)),]),
      x=1:ncol(lig16Rmat),y=1:nrow(lig16Rmat),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(10,palette="inferno",rev=T)[c(1,9)])
mtext(colnames(lig16Rmat),side=3,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
mtext(colnames(lig16Rmat),side=1,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
title("Ligand interaction with receptors",line=4.5)

par(mar=c(6,0,6,0.5),mgp=2:0,las=2)
image(z=t(meanR[rev(rownames(meanR)),]),
      x=1:ncol(meanR),y=1:nrow(meanR),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(100,palette="inferno",rev=T))
mtext(ct14[colnames(meanR)],side=3,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(ct14[colnames(meanR)],side=1,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(rev(rownames(meanR)),xpd=NA,side=2,line=0.2,at=1:nrow(meanR),cex=0.7,
      font=rev(as.integer(lvl3_ctl@rdesc[temp_genes,"pr_is_lm"]) + 1))

par(mar=c(0,3.5,6,0.5),mgp=2:0,las=0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",axes=F,
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",side=1,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```

```{r rec_per_lig_ccinx,fig.height=6,fig.width=9}
maxRL <- sapply(lig16R,function(X) apply(meanR[X,],2,max))
meanRL <- sapply(lig16R,function(X) apply(meanR[X,],2,mean))


maxRL <- maxRL[hclust(dist(maxRL),method="ward.D2")$order,
               hclust(dist(t(maxRL)),method="ward.D2")$order]
lig16Rmat <- lig16Rmat[,colnames(maxRL)]

layout(rbind(1:3),widths=c(1,3,2))
par(mar=c(9,3.5,2,1),mgp=2:0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",line=2.5,side=1)


par(mar=c(9,1,2,0))
image(z=maxRL,
      x=1:nrow(maxRL),y=1:ncol(maxRL),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      breaks=seq(0,15,length.out=1001),
      col=sequential_hcl(1000,palette="inferno",rev=T))
mtext("Max cognate receptor expression",font=2,line=0.5)
text(x=1:nrow(maxRL),y=rep(0,nrow(maxRL)),
     labels=rownames(maxRL),xpd=NA,srt=315,adj=0)

par(mar=c(9,5.5,2,1))
temp <- barplot(colSums(lig16Rmat),yaxs="i",axisnames=F,horiz=T)
mtext("Cognate receptors",line=2.5,side=1)
mtext(colnames(lig16Rmat),at=temp,
      side=2,line=2.5,adj=0.5,las=2)
```

## RF accuracy explained by receptor availability?

```{r load_data_lvl5_ccinx, message=FALSE, warning=FALSE, include=FALSE}
if (exists("lvl5_data")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs.RData") 
} else {
  source("lvl5_inputs.R")
}
```

```{r RF_leave1out_ccinx}
if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/200524_lvl5_mixall_balanced_leave1out_results.RData")) {
  temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/200524_lvl5_mixall_balanced_leave1out_results.RData")
} else {
  source("200524_lvl5_mixall_leave1out.R")
}

test_results <- data.frame(
  IDs=as.vector(testIDs),
  Labels=as.vector(t(sapply(rfresults,function(X) X$predictions))),
  stringsAsFactors=F
)
test_hits <- vector("list",sum(lvl5_data@cdesc$pert_iname %in% lig16))
names(test_hits) <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname %in% lig16]

for (X in 1:nrow(test_results)) {
  test_hits[[test_results[X,"IDs"]]] <- append(test_hits[[test_results[X,"IDs"]]],
                                               test_results[X,"Labels"])
}

acc_leave1out <- sapply(names(test_hits),function(X) 
  sum(test_hits[[X]] == lvl5_data@cdesc[X,"pert_iname"]) / 
    length(test_hits[[X]]))
acc_leave1out_totals <- sapply(lig16,function(X) {
  Y <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X]
  return(sum(unlist(test_hits[Y]) == X) / 
           length(unlist(test_hits[Y])))
})

acc_leave1out_ct <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    Z <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X &
                                     lvl5_data@cdesc$cell_id == Y]
    return(sum(unlist(test_hits[Z]) == X) / 
             length(unlist(test_hits[Z])))
  })
})

test_counts <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    sum(lvl5_data@cdesc$pert_iname == X &
          lvl5_data@cdesc$cell_id == Y)
  })
})

test_predictions <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    temp <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X &
                                        lvl5_data@cdesc$cell_id == Y]
    return(table(factor(unlist(test_hits[temp]),levels=lig16)) / 
             length(unlist(test_hits[temp])))
  })
},simplify=F)

rm(X,temp)
```

Cognate receptor expression for each ligand was compared to the accuracy per cell line of the leave-one-out ligand classification model outlined [previously](https://baderlab.github.io/Brendan_CCInxPred/LigPred_lvl5.html#cell-type-biases-in-accuracy).  Lines for each receptor are coloured by the spearman correlation coefficient between expression and accuracy per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.

```{r rec_expr_acc_corr_ccinx,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,xlim=range(acc_leave1out_ct[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
       xlab="Prediction accuracy per cell type",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  # segments(x0=seq(0,1,length.out=1000),
  #          x1=seq(0,1,length.out=1000),
  #          y0=rep(par("usr")[3],1000),
  #          y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
  #          lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))
  
  temp_order <- rownames(acc_leave1out_ct)[order(acc_leave1out_ct[,LIG])]
  temp_corr <- apply(meanR[lig16R[[LIG]],temp_order],1,function(X)
    cor(acc_leave1out_ct[temp_order,LIG],X,method="spearman"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(temp_corr)]) {
    points(acc_leave1out_ct[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(acc_leave1out_ct[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(acc_leave1out_ct[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r legend1,fig.height=0.8,fig.width=9}
par(mar=c(2.2,8,0.6,8))
barplot(rep(1,1000),border=NA,space=0,yaxt="n",xaxs="i",
        col=diverge_hcl(1000,palette="Purple-Green"))
mtext(c(-1,0,1),side=1,line=0.1,at=c(0,500,1000))
mtext("Spearman correlation coefficient",line=1.1,side=1,font=2)
```

## Number of DE genes explained by receptor availability?

```{r lig_DE_ccinx,fig.height=6,fig.width=9}
temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/lig16_DE_ligct_FDR.RData")

scoreDE <- pDE_ligct[,"01"] + 1e-4
scoreDE[scoreDE > 1] <- 1
scoreDE <- -log10(scoreDE)
  
scoreMAT <- tapply(scoreDE,as.factor(sub("^.+_","",names(scoreDE))),c)
for (X in names(scoreMAT)) {
  names(scoreMAT[[X]]) <- sub(paste0("_",X),"",names(scoreMAT[[X]]))
  scoreMAT[[X]] <- scoreMAT[[X]][sort(names(scoreMAT[[X]]))]
}
scoreMAT <- t(do.call(rbind,scoreMAT))
scoreMAT <- scoreMAT[rownames(maxRL),colnames(maxRL)]


layout(matrix(c(1,1,2,3),nrow=2),widths=c(2,1))

par(mar=c(3,2,7,6),mgp=2:0,las=2)
image(scoreMAT,
      col=sequential_hcl(100,palette="PinkYl",rev=T),
      x=1:nrow(scoreMAT),y=1:ncol(scoreMAT),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA)
mtext(colnames(scoreMAT),side=4,line=0.1,
      at=1:ncol(scoreMAT),adj=0,cex=switch((ncol(scoreMAT) > 20) + 1,0.9,0.5),
      col=qualitative_hcl(ncol(scoreMAT),palette="dark3"))
mtext("Ligands",side=4,line=4.5,las=0,font=2,cex=1.5,
      at=par("usr")[3] + (par("usr")[4] - par("usr")[3]) / 2,adj=0.5)
text(x=1:nrow(scoreMAT),y=rep(line2user(0.3,3),nrow(scoreMAT)),
     col=qualitative_hcl(nrow(scoreMAT),palette="dark3"),
     labels=rownames(scoreMAT),xpd=NA,adj=0,srt=45,cex=0.8)
text(x=0,y=line2user(1,3),labels="Cell lines",
     xpd=NA,adj=0,srt=45,font=2,cex=1.5)

rect(xleft=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,1] - 0.5,
     xright=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,1] + 0.5,
     ybottom=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,2] - 0.5,
     ytop=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,2] + 0.5,
     border="dodgerblue")

temp_left <- par("usr")[1] + (par("usr")[2] - par("usr")[1]) * 0.3
temp_right <- par("usr")[1] + (par("usr")[2] - par("usr")[1]) * 0.7

segments(x0=seq(temp_left,temp_right,length.out=1000),
         x1=seq(temp_left,temp_right,length.out=1000),
         y0=rep(line2user(1,1),1000),y1=rep(line2user(1.6,1),1000),
         xpd=NA,col=sequential_hcl(1000,palette="PinkYl",rev=T))
mtext(c("100%","< 0.01%"),side=1,line=0.8,las=0,adj=c(1.1,-0.1),
      at=c(temp_left,temp_right))
mtext("Probability of at least # DE occuring by chance",
      las=0,side=1,line=1.7,adj=0.5,
      at=temp_left + (temp_right - temp_left) / 2)

rect(xleft=temp_left + (-log10(0.05 + 1e-4) / max(scoreMAT)) * (temp_right - temp_left),
     xright=temp_right,ybottom=line2user(1,1),ytop=line2user(1.6,1),
     xpd=NA,col=NA,border="dodgerblue")
mtext("p <= 0.05",side=1,las=0,col="dodgerblue",cex=0.9,line=0,adj=1,at=temp_right)

rm(list=temp)
rm(list=c("X",grep("^temp",ls(),value=T)))



par(mar=c(3,3,2,1),mgp=2:0,las=0)
plot(as.vector(maxRL),as.vector(scoreMAT),pch=20,
     col=rep(qualitative_hcl(nrow(maxRL),palette="dark3"),ncol(maxRL)), # per cell line
     xlab="Max cognate receptor expression",
     ylab="# DE probability score")
mtext("Coloured by cell line",font=2,line=0.1)
mtext(paste("PCC =",signif(cor(as.vector(maxRL),as.vector(scoreMAT)),2)),
      line=-1,adj=0.05,cex=0.9)
plot(as.vector(maxRL),as.vector(scoreMAT),pch=20,
     col=as.vector(sapply(qualitative_hcl(ncol(maxRL),palette="dark3"),rep,times=nrow(maxRL))), # per ligand
     xlab="Max cognate receptor expression",
     ylab="# DE probability score")
mtext("Coloured by ligand",font=2,line=0.1)
mtext(paste("PCC =",signif(cor(as.vector(maxRL),as.vector(scoreMAT)),2)),
      line=-1,adj=0.05,cex=0.9)

```

Cognate receptor expression for each ligand was compared to probability scores for the number of differentially expressed genes associated with each ligand-treated cell line outlined [previously](https://baderlab.github.io/Brendan_CCInxPred/DEoverlap_FDR.html#averaging-within-each-ligand-per-cell-line).  Lines for each receptor are coloured by the spearman correlation coefficient between expression and DE score per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.

```{r rec_DE_corr_ccinx}
Rcor_scoreDE <- sapply(lig16,function(LIG)
  apply(meanR[lig16R[[LIG]],rownames(scoreMAT)],1,function(X) 
    cor(X,scoreMAT[,LIG],method="spearman")))

```



```{r rec_DE_corr_detail_ccinx,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,xlim=range(scoreMAT[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
       xlab="# DE probability score",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  # segments(x0=seq(min(scoreMAT),max(scoreMAT),length.out=1000),
  #          x1=seq(min(scoreMAT),max(scoreMAT),length.out=1000),
  #          y0=rep(par("usr")[3],1000),
  #          y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
  #          lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))

  temp_order <- rownames(scoreMAT)[order(scoreMAT[,LIG])]
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,Rcor_scoreDE[[LIG]]),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(Rcor_scoreDE[[LIG]])]) {
    points(scoreMAT[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(Rcor_scoreDE[[LIG]])[order(Rcor_scoreDE[[LIG]],decreasing=T)][1:3]
  text(scoreMAT[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(scoreMAT[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```
```{r legend2,fig.height=0.8,fig.width=9}
par(mar=c(2.2,8,0.6,8))
barplot(rep(1,1000),border=NA,space=0,yaxt="n",xaxs="i",
        col=diverge_hcl(1000,palette="Purple-Green"))
mtext(c(-1,0,1),side=1,line=0.1,at=c(0,500,1000))
mtext("Spearman correlation coefficient",line=1.1,side=1,font=2)
```

```{r cleanup1}
rm(list=ls())
source("~/Dropbox/GDB/line2user.R")
```


****

# Receptor mapping from FANTOM5

Ligand-receptor interactions are from the FANTOM5 database (Ramilowski et al.)

## Receptor expression per cell line  

```{r load_data_lvl3_ctl_FANTOM5,include=TRUE}
if (exists("lvl3_ctl")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData") 
} else {
  source("lvl3_inputs_allgenesctrl.R")
}

# temp <- load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_human.RData",package="CCInx"))
temp <- load(system.file("LigRecDB_RData/FANTOM5_human.RData",package="CCInx"))

lig16R <- sapply(lig16,function(X) {
  temp <- c(inxDB$nodeB[inxDB$nodeA == X],inxDB$nodeA[inxDB$nodeB == X])
  return(temp[grepl("Receptor",geneInfo[temp,"protein_type"])])
},simplify=F)
# lig16R[["GAS6"]] <- c("AXL","TYRO3","MERTK")
for (LIG in lig16) {
  temp_in <- lig16R[[LIG]] %in% lvl3_ctl@rdesc$pr_gene_symbol
  lig16R[[LIG]] <- lig16R[[LIG]][temp_in]
  names(lig16R[[LIG]]) <- sapply(lig16R[[LIG]],function(X) 
    lvl3_ctl@rdesc$id[lvl3_ctl@rdesc$pr_gene_symbol == X])
  cat(paste0(round(sum(temp_in) / length(temp_in) * 100),"% (",
               round(sum(lvl3_ctl@rdesc[names(lig16R[[LIG]]),"pr_is_lm"] == "1") / length(temp_in) * 100),
               "%) of ",LIG," receptors present in CMap data, either inferred or (assayed)."),"\n")
}

rm(list=c("LIG","temp",temp))
```

Below is an overview of all potential cognate receptors per ligand present in the data.  The left figure shows which ligands (dark purple) each receptor is known to interact with.  On the right is a heatmap showing mean normalized gene expression per cell type for each receptor gene, with darker indicating higher expression.  Gene names (rows) are ordered by heirarchical clustering of ligand-receptor interaction data, such that receptors common to the same ligand should be together.  **Landmark** gene expression was emperically determined rather than inferred (names indicated in bold).  Ligand names (columns, left) are ordered by heirarchical clustering of ligand-receptor interaction data, such that ligands sharing the same receptors should be proximal.  Cell type names (columns, right) are ordered by heirarchical clustering of mean receptor gene expression.

```{r rec_expr_FANTOM5,fig.height=9,fig.width=9}
temp_genes <- unique(unlist(lapply(lig16R,names)))
names(temp_genes) <- lvl3_ctl@rdesc[temp_genes,"pr_gene_symbol"]
lig16Rmat <- sapply(lig16R,function(X) names(temp_genes) %in% X)
rownames(lig16Rmat) <- names(temp_genes)
temp_roworder <- hclust(dist(lig16Rmat))$order
lig16Rmat <- lig16Rmat[temp_roworder,]
lig16Rmat <- lig16Rmat[,hclust(dist(t(lig16Rmat)))$order]

meanR <- sapply(ct14,function(CT) rowMeans(lvl3_ctl@mat[temp_genes,lvl3_ctl@cdesc$cell_id == CT]))
rownames(meanR) <- names(temp_genes)
meanR <- meanR[temp_roworder,]
meanR <- meanR[,hclust(dist(t(meanR)))$order]

layout(cbind(rep(1,25),rep(2,25),c(rep(3,5),rep(0,20))),widths=c(3,3,1))
par(mar=c(6,0.5,6,5.5),mgp=2:0,las=2)
image(z=t(lig16Rmat[rev(rownames(lig16Rmat)),]),
      x=1:ncol(lig16Rmat),y=1:nrow(lig16Rmat),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(10,palette="inferno",rev=T)[c(1,9)])
mtext(colnames(lig16Rmat),side=3,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
mtext(colnames(lig16Rmat),side=1,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
title("Ligand interaction with receptors",line=4.5)

par(mar=c(6,0,6,0.5),mgp=2:0,las=2)
image(z=t(meanR[rev(rownames(meanR)),]),
      x=1:ncol(meanR),y=1:nrow(meanR),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(100,palette="inferno",rev=T))
mtext(ct14[colnames(meanR)],side=3,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(ct14[colnames(meanR)],side=1,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(rev(rownames(meanR)),xpd=NA,side=2,line=0.2,at=1:nrow(meanR),cex=0.7,
      font=rev(as.integer(lvl3_ctl@rdesc[temp_genes,"pr_is_lm"]) + 1))

par(mar=c(0,3.5,6,0.5),mgp=2:0,las=0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",axes=F,
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",side=1,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```

```{r rec_per_lig_FANTOM5,fig.height=6,fig.width=9}
maxRL <- sapply(lig16R,function(X) apply(meanR[X,],2,max))
meanRL <- sapply(lig16R,function(X) apply(meanR[X,],2,mean))


maxRL <- maxRL[hclust(dist(maxRL),method="ward.D2")$order,
               hclust(dist(t(maxRL)),method="ward.D2")$order]
lig16Rmat <- lig16Rmat[,colnames(maxRL)]

layout(rbind(1:3),widths=c(1,3,2))
par(mar=c(9,3.5,2,1),mgp=2:0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",line=2.5,side=1)


par(mar=c(9,1,2,0))
image(z=maxRL,
      x=1:nrow(maxRL),y=1:ncol(maxRL),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      breaks=seq(0,15,length.out=1001),
      col=sequential_hcl(1000,palette="inferno",rev=T))
mtext("Max cognate receptor expression",font=2,line=0.5)
text(x=1:nrow(maxRL),y=rep(0,nrow(maxRL)),
     labels=rownames(maxRL),xpd=NA,srt=315,adj=0)

par(mar=c(9,5.5,2,1))
temp <- barplot(colSums(lig16Rmat),yaxs="i",axisnames=F,horiz=T)
mtext("Cognate receptors",line=2.5,side=1)
mtext(colnames(lig16Rmat),at=temp,
      side=2,line=2.5,adj=0.5,las=2)
```

## RF accuracy explained by receptor availability?

```{r load_data_FANTOM5, message=FALSE, warning=FALSE, include=FALSE}
if (exists("lvl5_data")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs.RData") 
} else {
  source("lvl5_inputs.R")
}
```

```{r RF_leave1out_FANTOM5}
if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/200524_lvl5_mixall_balanced_leave1out_results.RData")) {
  temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/200524_lvl5_mixall_balanced_leave1out_results.RData")
} else {
  source("200524_lvl5_mixall_leave1out.R")
}

test_results <- data.frame(
  IDs=as.vector(testIDs),
  Labels=as.vector(t(sapply(rfresults,function(X) X$predictions))),
  stringsAsFactors=F
)
test_hits <- vector("list",sum(lvl5_data@cdesc$pert_iname %in% lig16))
names(test_hits) <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname %in% lig16]

for (X in 1:nrow(test_results)) {
  test_hits[[test_results[X,"IDs"]]] <- append(test_hits[[test_results[X,"IDs"]]],
                                               test_results[X,"Labels"])
}

acc_leave1out <- sapply(names(test_hits),function(X) 
  sum(test_hits[[X]] == lvl5_data@cdesc[X,"pert_iname"]) / 
    length(test_hits[[X]]))
acc_leave1out_totals <- sapply(lig16,function(X) {
  Y <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X]
  return(sum(unlist(test_hits[Y]) == X) / 
           length(unlist(test_hits[Y])))
})

acc_leave1out_ct <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    Z <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X &
                                     lvl5_data@cdesc$cell_id == Y]
    return(sum(unlist(test_hits[Z]) == X) / 
             length(unlist(test_hits[Z])))
  })
})

test_counts <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    sum(lvl5_data@cdesc$pert_iname == X &
          lvl5_data@cdesc$cell_id == Y)
  })
})

test_predictions <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    temp <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X &
                                        lvl5_data@cdesc$cell_id == Y]
    return(table(factor(unlist(test_hits[temp]),levels=lig16)) / 
             length(unlist(test_hits[temp])))
  })
},simplify=F)

rm(X,temp)
```

Cognate receptor expression for each ligand was compared to the accuracy per cell line of the leave-one-out ligand classification model outlined [previously](https://baderlab.github.io/Brendan_CCInxPred/LigPred_lvl5.html#cell-type-biases-in-accuracy).  Lines for each receptor are coloured by the spearman correlation coefficient between expression and accuracy per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.

```{r rec_expr_acc_corr_FANTOM5,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,xlim=range(acc_leave1out_ct[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
       xlab="Prediction accuracy per cell type",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  # segments(x0=seq(0,1,length.out=1000),
  #          x1=seq(0,1,length.out=1000),
  #          y0=rep(par("usr")[3],1000),
  #          y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
  #          lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))
  
  temp_order <- rownames(acc_leave1out_ct)[order(acc_leave1out_ct[,LIG])]
  temp_corr <- apply(meanR[lig16R[[LIG]],temp_order],1,function(X)
    cor(acc_leave1out_ct[temp_order,LIG],X,method="spearman"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(temp_corr)]) {
    points(acc_leave1out_ct[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(acc_leave1out_ct[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(acc_leave1out_ct[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r legend3,fig.height=0.8,fig.width=9}
par(mar=c(2.2,8,0.6,8))
barplot(rep(1,1000),border=NA,space=0,yaxt="n",xaxs="i",
        col=diverge_hcl(1000,palette="Purple-Green"))
mtext(c(-1,0,1),side=1,line=0.1,at=c(0,500,1000))
mtext("Spearman correlation coefficient",line=1.1,side=1,font=2)
```

## Number of DE genes explained by receptor availability?

```{r lig_DE_FANTOM5,fig.height=6,fig.width=9}
temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/lig16_DE_ligct_FDR.RData")

scoreDE <- pDE_ligct[,"01"] + 1e-4
scoreDE[scoreDE > 1] <- 1
scoreDE <- -log10(scoreDE)
  
scoreMAT <- tapply(scoreDE,as.factor(sub("^.+_","",names(scoreDE))),c)
for (X in names(scoreMAT)) {
  names(scoreMAT[[X]]) <- sub(paste0("_",X),"",names(scoreMAT[[X]]))
  scoreMAT[[X]] <- scoreMAT[[X]][sort(names(scoreMAT[[X]]))]
}
scoreMAT <- t(do.call(rbind,scoreMAT))
scoreMAT <- scoreMAT[rownames(maxRL),colnames(maxRL)]


layout(matrix(c(1,1,2,3),nrow=2),widths=c(2,1))

par(mar=c(3,2,7,6),mgp=2:0,las=2)
image(scoreMAT,
      col=sequential_hcl(100,palette="PinkYl",rev=T),
      x=1:nrow(scoreMAT),y=1:ncol(scoreMAT),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA)
mtext(colnames(scoreMAT),side=4,line=0.1,
      at=1:ncol(scoreMAT),adj=0,cex=switch((ncol(scoreMAT) > 20) + 1,0.9,0.5),
      col=qualitative_hcl(ncol(scoreMAT),palette="dark3"))
mtext("Ligands",side=4,line=4.5,las=0,font=2,cex=1.5,
      at=par("usr")[3] + (par("usr")[4] - par("usr")[3]) / 2,adj=0.5)
text(x=1:nrow(scoreMAT),y=rep(line2user(0.3,3),nrow(scoreMAT)),
     col=qualitative_hcl(nrow(scoreMAT),palette="dark3"),
     labels=rownames(scoreMAT),xpd=NA,adj=0,srt=45,cex=0.8)
text(x=0,y=line2user(1,3),labels="Cell lines",
     xpd=NA,adj=0,srt=45,font=2,cex=1.5)

rect(xleft=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,1] - 0.5,
     xright=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,1] + 0.5,
     ybottom=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,2] - 0.5,
     ytop=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,2] + 0.5,
     border="dodgerblue")

temp_left <- par("usr")[1] + (par("usr")[2] - par("usr")[1]) * 0.3
temp_right <- par("usr")[1] + (par("usr")[2] - par("usr")[1]) * 0.7

segments(x0=seq(temp_left,temp_right,length.out=1000),
         x1=seq(temp_left,temp_right,length.out=1000),
         y0=rep(line2user(1,1),1000),y1=rep(line2user(1.6,1),1000),
         xpd=NA,col=sequential_hcl(1000,palette="PinkYl",rev=T))
mtext(c("100%","< 0.01%"),side=1,line=0.8,las=0,adj=c(1.1,-0.1),
      at=c(temp_left,temp_right))
mtext("Probability of at least # DE occuring by chance",
      las=0,side=1,line=1.7,adj=0.5,
      at=temp_left + (temp_right - temp_left) / 2)

rect(xleft=temp_left + (-log10(0.05 + 1e-4) / max(scoreMAT)) * (temp_right - temp_left),
     xright=temp_right,ybottom=line2user(1,1),ytop=line2user(1.6,1),
     xpd=NA,col=NA,border="dodgerblue")
mtext("p <= 0.05",side=1,las=0,col="dodgerblue",cex=0.9,line=0,adj=1,at=temp_right)

rm(list=temp)
rm(list=c("X",grep("^temp",ls(),value=T)))



par(mar=c(3,3,2,1),mgp=2:0,las=0)
plot(as.vector(maxRL),as.vector(scoreMAT),pch=20,
     col=rep(qualitative_hcl(nrow(maxRL),palette="dark3"),ncol(maxRL)), # per cell line
     xlab="Max cognate receptor expression",
     ylab="# DE probability score")
mtext("Coloured by cell line",font=2,line=0.1)
mtext(paste("PCC =",signif(cor(as.vector(maxRL),as.vector(scoreMAT)),2)),
      line=-1,adj=0.05,cex=0.9)
plot(as.vector(maxRL),as.vector(scoreMAT),pch=20,
     col=as.vector(sapply(qualitative_hcl(ncol(maxRL),palette="dark3"),rep,times=nrow(maxRL))), # per ligand
     xlab="Max cognate receptor expression",
     ylab="# DE probability score")
mtext("Coloured by ligand",font=2,line=0.1)
mtext(paste("PCC =",signif(cor(as.vector(maxRL),as.vector(scoreMAT)),2)),
      line=-1,adj=0.05,cex=0.9)

```

Cognate receptor expression for each ligand was compared to probability scores for the number of differentially expressed genes associated with each ligand-treated cell line outlined [previously](https://baderlab.github.io/Brendan_CCInxPred/DEoverlap_FDR.html#averaging-within-each-ligand-per-cell-line).  Lines for each receptor are coloured by the spearman correlation coefficient between expression and DE score per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.

```{r rec_DE_corr_FANTOM5}
Rcor_scoreDE <- sapply(lig16,function(LIG)
  apply(meanR[lig16R[[LIG]],rownames(scoreMAT)],1,function(X) 
    cor(X,scoreMAT[,LIG],method="spearman")))

```



```{r rec_DE_corr_detail_FANTOM5,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,xlim=range(scoreMAT[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
       xlab="# DE probability score",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  # segments(x0=seq(min(scoreMAT),max(scoreMAT),length.out=1000),
  #          x1=seq(min(scoreMAT),max(scoreMAT),length.out=1000),
  #          y0=rep(par("usr")[3],1000),
  #          y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
  #          lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))
  
  temp_order <- rownames(scoreMAT)[order(scoreMAT[,LIG])]
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,Rcor_scoreDE[[LIG]]),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(Rcor_scoreDE[[LIG]])]) {
    points(scoreMAT[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(Rcor_scoreDE[[LIG]])[order(Rcor_scoreDE[[LIG]],decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(scoreMAT[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(scoreMAT[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

## Transcriptional Activity Score explained by receptor availability?

```{r loadTAS}
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_TAS.RData")

meanTAS <- sapply(lig16,function(LIG) {
  sapply(ct14,function(CT) {
    mean(TASdf$TAS[TASdf$cell_id == CT & TASdf$pert_iname == LIG])
  })
})
meanTAS <- meanTAS[colnames(meanR),]

```


```{r TAS_acc_pDE_corr,fig.height=4,fig.width=8}
listTAS <- sapply(lig16,function(LIG) {
    sapply(ct14,function(CT) {
        TASdf$TAS[TASdf$cell_id == CT & TASdf$pert_iname == LIG]
    })
},simplify=F)

temp_acc <- sapply(colnames(acc_leave1out_ct),function(LIG) {
  sapply(rownames(acc_leave1out_ct),function(CT) {
    rep(acc_leave1out_ct[CT,LIG],length(listTAS[[LIG]][[CT]]))
  })
},simplify=F)

temp_de <- sapply(colnames(scoreMAT),function(LIG) {
  sapply(rownames(scoreMAT),function(CT) {
    rep(scoreMAT[CT,LIG],length(listTAS[[LIG]][[CT]]))
  })
},simplify=F)

par(mar=c(3,3,2,1),mgp=2:0,mfrow=c(1,2))
plot(unlist(temp_acc),unlist(listTAS),pch=".",cex=2,
     xlab="RF Acc",ylab="TAS")
mtext(paste("PCC:",signif(cor(unlist(temp_acc),unlist(listTAS)),2)))
plot(unlist(temp_de),unlist(listTAS),pch=".",cex=2,
     xlab="scoreDE",ylab="TAS")
mtext(paste("PCC:",signif(cor(unlist(temp_de),unlist(listTAS)),2)))

```

```{r rec_expr_TAS_corr_FANTOM5,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,ylim=c(0,15),bty="n",yaxt="n",
       xlim=range(meanTAS[,LIG]),# xlim=c(0,1),
       xlab="Mean TAS per cell type",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  # segments(x0=seq(0,1,length.out=1000),
  #          x1=seq(0,1,length.out=1000),
  #          y0=rep(par("usr")[3],1000),
  #          y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
  #          lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))
  
  temp_order <- rownames(meanTAS)[order(meanTAS[,LIG])]
  temp_corr <- apply(meanR[lig16R[[LIG]],temp_order],1,function(X)
    cor(meanTAS[temp_order,LIG],X,method="spearman"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(temp_corr)]) {
    points(meanTAS[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(meanTAS[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(meanTAS[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r legend5,fig.height=0.8,fig.width=9}
par(mar=c(2.2,8,0.6,8))
barplot(rep(1,1000),border=NA,space=0,yaxt="n",xaxs="i",
        col=diverge_hcl(1000,palette="Purple-Green"))
mtext(c(-1,0,1),side=1,line=0.1,at=c(0,500,1000))
mtext("Spearman correlation coefficient",line=1.1,side=1,font=2)
```




****

# Proteomics from DepMap, receptor mapping from FANTOM5

Ligand-receptor interactions are from the FANTOM5 database (Ramilowski et al.)

## Receptor protein abundance per cell line 

```{r load_data_depmap,include=TRUE}
rm(list=ls()[!ls() %in% c("meanR","acc_leave1out_ct","scoreMAT","meanTAS","lig16R","ct14","lig16")])

temp_MS <- read.csv("~/Dropbox/GDB_archive/CMapCorr_files/protein_quant_current_normalized.csv.gz")
temp_receptors <- temp_MS$Gene_Symbol %in% unique(unlist(lig16R))
temp_lines <- unlist(sapply(ct14,function(X) 
                grep(paste0("^",X,"_"),colnames(temp_MS),value=T))) 
temp_all_MS <- temp_MS[temp_receptors,-c(1:6)]
temp_all_MS <- temp_all_MS[,!grepl("^TenP",colnames(temp_all_MS))]
temp_z_MS <- apply(temp_all_MS,1,function(X) (X - mean(X,na.rm=T)) / sd(X,na.rm=T))

MS <- t(temp_z_MS[temp_lines,])
rownames(MS) <- temp_MS$Gene_Symbol[temp_receptors]
colnames(MS) <- names(temp_lines)
MS <- MS[!apply(MS,1,function(X) all(is.na(X))),]

lig16R_MS <- sapply(lig16R,function(X) X[X %in% rownames(MS)])

rm(list=grep("^temp",ls(),value=T))
```


```{r rec_expr_TAS_corr_depmap,fig.height=9,fig.width=9,fig.show='hold'}
temp_TAS <- meanTAS[colnames(MS),]

par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,bty="n",yaxt="n",
       ylim=range(MS[lig16R_MS[[LIG]],],na.rm=T),
       xlim=range(temp_TAS[,LIG]),# xlim=c(0,1),
       xlab="Mean TAS per cell type",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor relative abundance per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)

  temp_order <- rownames(temp_TAS)[order(temp_TAS[,LIG])]
  temp_corr <- apply(MS[lig16R_MS[[LIG]],temp_order,drop=F],1,function(X)
    cor(temp_TAS[temp_order,LIG],X,method="spearman",use="na.or.complete"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R_MS[[LIG]]
  for (GENE in lig16R_MS[[LIG]][order(temp_corr)]) {
    temp_MS <- MS[GENE,temp_order]
    temp_MS <- temp_MS[!is.na(temp_MS)]
    if (length(temp_MS) < 1) { next }
    points(temp_TAS[names(temp_MS),LIG],temp_MS,
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(temp_TAS[temp_order[length(temp_order)],LIG],
       MS[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4)
  text(temp_TAS[temp_order[1],LIG],
       MS[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r legend6,fig.height=0.8,fig.width=9}
par(mar=c(2.2,8,0.6,8))
barplot(rep(1,1000),border=NA,space=0,yaxt="n",xaxs="i",
        col=diverge_hcl(1000,palette="Purple-Green"))
mtext(c(-1,0,1),side=1,line=0.1,at=c(0,500,1000))
mtext("Spearman correlation coefficient",line=1.1,side=1,font=2)
```

```{r rec_expr_acc_corr_depmap,fig.height=9,fig.width=9,fig.show='hold'}
temp_acc <- acc_leave1out_ct[colnames(MS),]

par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,bty="n",yaxt="n",
       ylim=range(MS[lig16R_MS[[LIG]],],na.rm=T),
       xlim=range(temp_acc[,LIG]),# xlim=c(0,1),
       xlab="Prediction accuracy per cell type",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor relative abundance per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)

  temp_order <- rownames(temp_acc)[order(temp_acc[,LIG])]
  temp_corr <- apply(MS[lig16R_MS[[LIG]],temp_order,drop=F],1,function(X)
    cor(temp_acc[temp_order,LIG],X,method="spearman",use="na.or.complete"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R_MS[[LIG]]
  for (GENE in lig16R_MS[[LIG]][order(temp_corr)]) {
    temp_MS <- MS[GENE,temp_order]
    temp_MS <- temp_MS[!is.na(temp_MS)]
    if (length(temp_MS) < 1) { next }
    points(temp_acc[names(temp_MS),LIG],temp_MS,
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(temp_acc[temp_order[length(temp_order)],LIG],
       MS[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4)
  text(temp_acc[temp_order[1],LIG],
       MS[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r rec_expr_DE_corr_depmap, fig.height=9,fig.width=9}
temp_scoreMAT <- scoreMAT[colnames(MS),]

par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,bty="n",yaxt="n",
       ylim=range(MS[lig16R_MS[[LIG]],],na.rm=T),
       xlim=range(temp_scoreMAT[,LIG]),# xlim=c(0,1),
       xlab="# DE probability score",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor relative abundance per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)

  temp_order <- rownames(temp_scoreMAT)[order(temp_scoreMAT[,LIG])]
  temp_corr <- apply(MS[lig16R_MS[[LIG]],temp_order,drop=F],1,function(X)
    cor(temp_scoreMAT[temp_order,LIG],X,method="spearman",use="na.or.complete"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R_MS[[LIG]]
  for (GENE in lig16R_MS[[LIG]][order(temp_corr)]) {
    temp_MS <- MS[GENE,temp_order]
    temp_MS <- temp_MS[!is.na(temp_MS)]
    if (length(temp_MS) < 1) { next }
    points(temp_scoreMAT[names(temp_MS),LIG],temp_MS,
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  temp_top <- temp_top[!is.na(temp_top)]
  text(temp_scoreMAT[temp_order[length(temp_order)],LIG],
       MS[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4)
  text(temp_scoreMAT[temp_order[1],LIG],
       MS[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))

```

## Correlation between mRNA and protein abundance

```{r expr_prot_corr, fig.height=5,fig.width=6}
temp_meanR <- as.vector(meanR[rownames(MS),colnames(MS)])
temp_MS <- as.vector(MS)

par(mar=c(3,3,2,1),mgp=2:0)
plot(temp_meanR,temp_MS,pch=20, #pch=".",cex=2,
     xlab="Receptor normalized expression per cell type",
     ylab="Receptor relative abundance per cell type")
mtext(paste("Spearman rank correlation:",
            signif(cor(temp_meanR,temp_MS,method="spearman",use="na.or.complete"),3)))

rm(list=grep("^temp",ls(),value=T))
```


# Summary figures

```{r summary, fig.width=8,fig.height=12}
SUMMPLOT <- function(XDAT,YDAT,XLAB,YLAB) {
  temp_lig16R <- sapply(lig16R,function(X) X[X %in% rownames(YDAT)])
  temp_X <- XDAT[colnames(YDAT),]
  
  temp_R <- unlist(sapply(rownames(temp_X),function(CT) {
    sapply(colnames(temp_X),function(LIG) {
      YDAT[temp_lig16R[[LIG]],CT][!is.na(YDAT[temp_lig16R[[LIG]],CT])]
    })
  },simplify=F),recursive=F)
  
  plot(NA,NA,xlab=XLAB,ylab=YLAB,
       xlim=range(temp_X,na.rm=T),ylim=range(temp_R,na.rm=T))
  for (X in seq_along(temp_R)) {
    if (length(temp_R[[X]]) <= 0) { next }
    points(rep(as.vector(temp_X)[X],length(temp_R[[X]])),
           sort(temp_R[[X]],na.last=T),
           type="l",col=alpha(1,0.2))
    points(rep(as.vector(temp_X)[X],length(temp_R[[X]])),
           sort(temp_R[[X]]),
           pch=".",cex=3)
  }
}

par(mfcol=c(3,2),mar=c(3,3,2,1),mgp=2:0)
SUMMPLOT(acc_leave1out_ct,meanR,"Random forest accuracy","Receptor normalized gene expression")
SUMMPLOT(scoreMAT,meanR,"#DE probability score","Receptor normalized gene expression")
SUMMPLOT(meanTAS,meanR,"Transcriptional activity score","Receptor normalized gene expression")
SUMMPLOT(acc_leave1out_ct,MS,"Random forest accuracy","Receptor relative protein abundance")
SUMMPLOT(scoreMAT,MS,"#DE probability score","Receptor relative protein abundance")
SUMMPLOT(meanTAS,MS,"Transcriptional activity score","Receptor relative protein abundance")

```

## 295 ligands in 9 cell lines

```{r prep295_lig}
rm(list=ls())

temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/lvl4_inputs.RData")

temp_ligcountsperct <- sapply(unique(lvl4_data@cdesc$cell_id),function(CT)
  length(unique(lvl4_data@cdesc[lvl4_data@cdesc$cell_id == CT,"pert_iname"])))
ct9 <- names(temp_ligcountsperct)[temp_ligcountsperct > 100]
names(ct9) <- sapply(ct9,function(X) names(ct14)[ct14 == X])

temp_ctcountsperlig <- sapply(unique(lvl4_data@cdesc$pert_iname),function(LIG)
  length(unique(lvl4_data@cdesc[lvl4_data@cdesc$pert_iname == LIG,"cell_id"])))

lig295 <- names(temp_ctcountsperlig)[temp_ctcountsperlig >= 9]

rm(list=c(temp,grep("^temp",ls(),value=T)))
```

```{r prep295_meanR}
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData") 
rm(lig16,ct14)
# temp <- load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_human.RData",package="CCInx"))
temp <- load(system.file("LigRecDB_RData/FANTOM5_human.RData",package="CCInx"))

lig295R <- sapply(lig295,function(X) {
  temp <- c(inxDB$nodeB[inxDB$nodeA == X],inxDB$nodeA[inxDB$nodeB == X])
  return(temp[grepl("Receptor",geneInfo[temp,"protein_type"])])
},simplify=F)
lig295R <- lig295R[sapply(lig295R,length) > 0]
rm(lig295)
for (LIG in names(lig295R)) {
  temp_in <- lig295R[[LIG]] %in% lvl3_ctl@rdesc$pr_gene_symbol
  lig295R[[LIG]] <- lig295R[[LIG]][temp_in]
  names(lig295R[[LIG]]) <- sapply(lig295R[[LIG]],function(X) 
    lvl3_ctl@rdesc$id[lvl3_ctl@rdesc$pr_gene_symbol == X])
}

temp_genes <- unique(unlist(lapply(lig295R,names)))
names(temp_genes) <- lvl3_ctl@rdesc[temp_genes,"pr_gene_symbol"]
meanR <- sapply(ct9,function(CT) rowMeans(lvl3_ctl@mat[temp_genes,lvl3_ctl@cdesc$cell_id == CT]))
rownames(meanR) <- names(temp_genes)

rm(list=c("LIG","lvl3_ctl",temp,grep("^temp",ls(),value=T)))
```

```{r prep295_protMS}
temp_MS <- read.csv("~/Dropbox/GDB_archive/CMapCorr_files/protein_quant_current_normalized.csv.gz")
temp_receptors <- temp_MS$Gene_Symbol %in% unique(unlist(lig295R))
temp_lines <- unlist(sapply(ct9,function(X) 
                grep(paste0("^",X,"_"),colnames(temp_MS),value=T))) 
temp_all_MS <- temp_MS[temp_receptors,-c(1:6)]
temp_all_MS <- temp_all_MS[,!grepl("^TenP",colnames(temp_all_MS))]
temp_z_MS <- apply(temp_all_MS,1,function(X) (X - mean(X,na.rm=T)) / sd(X,na.rm=T))

MS <- t(temp_z_MS[temp_lines,])
rownames(MS) <- temp_MS$Gene_Symbol[temp_receptors]
colnames(MS) <- names(temp_lines)
MS <- MS[!apply(MS,1,function(X) all(is.na(X))),]

rm(list=grep("^temp",ls(),value=T))
```

```{r 295_expr_prot_corr, fig.height=5,fig.width=6}
temp_meanR <- as.vector(meanR[rownames(MS),colnames(MS)])
temp_MS <- as.vector(MS)

par(mar=c(3,3,2,1),mgp=2:0)
plot(temp_meanR,temp_MS,pch=20,col=alpha(1,0.5), #pch=".",cex=2,
     xlab="Receptor normalized expression per cell type",
     ylab="Receptor relative abundance per cell type")
mtext(paste("Spearman rank correlation:",
            signif(cor(temp_meanR,temp_MS,method="spearman",use="na.or.complete"),3)))

rm(list=grep("^temp",ls(),value=T))
```


```{r prep295_scoreDE}
temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/lig295_DE_ligct_FDR.RData")

scoreDE <- pDE_ligct[,"01"] + 1e-4
scoreDE[scoreDE > 1] <- 1
scoreDE <- -log10(scoreDE)
  
scoreMAT <- tapply(scoreDE,as.factor(sub("^.+_","",names(scoreDE))),c)
for (X in names(scoreMAT)) {
  names(scoreMAT[[X]]) <- sub(paste0("_",X),"",names(scoreMAT[[X]]))
  scoreMAT[[X]] <- scoreMAT[[X]][sort(names(scoreMAT[[X]]))]
}
scoreMAT <- t(do.call(rbind,scoreMAT))

rm(list=c(temp,"temp","X","scoreDE"))
```

```{r prep295_TAS}
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_TAS.RData")
meanTAS <- sapply(names(lig295R),function(LIG) {
  sapply(ct9,function(CT) {
    mean(TASdf$TAS[TASdf$cell_id == CT & TASdf$pert_iname == LIG])
  })
})
rm(TASdf)
```

```{r summary295, fig.width=8,fig.height=8}
SUMMPLOT <- function(XDAT,YDAT,XLAB,YLAB) {
  temp_lig295R <- sapply(lig295R,function(X) X[X %in% rownames(YDAT)])
  temp_X <- XDAT[colnames(YDAT),]
  
  temp_R <- unlist(sapply(rownames(temp_X),function(CT) {
    sapply(colnames(temp_X),function(LIG) {
      YDAT[temp_lig295R[[LIG]],CT][!is.na(YDAT[temp_lig295R[[LIG]],CT])]
    })
  },simplify=F),recursive=F)
  
  plot(NA,NA,xlab=XLAB,ylab=YLAB,
       xlim=range(temp_X,na.rm=T),ylim=range(temp_R,na.rm=T))
  for (X in seq_along(temp_R)) {
    if (length(temp_R[[X]]) <= 0) { next }
    points(rep(as.vector(temp_X)[X],length(temp_R[[X]])),
           sort(temp_R[[X]],na.last=T),
           type="l",col=alpha(1,0.1))
    points(rep(as.vector(temp_X)[X],length(temp_R[[X]])),
           sort(temp_R[[X]]),
           pch=".",col=alpha(1,0.5),cex=3)
  }
}

par(mfcol=c(2,2),mar=c(3,3,2,1),mgp=2:0)
SUMMPLOT(scoreMAT,meanR,"#DE probability score","Receptor normalized gene expression")
SUMMPLOT(meanTAS,meanR,"Transcriptional activity score","Receptor normalized gene expression")
SUMMPLOT(scoreMAT,MS,"#DE probability score","Receptor relative protein abundance")
SUMMPLOT(meanTAS,MS,"Transcriptional activity score","Receptor relative protein abundance")

```