---
title: "Biological factors affecting prediction accuracy"
---
  
```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(cmapR)
library(ranger)
library(colorspace)
library(kableExtra)
```


# Receptor gene expression

The quantile-normalized transcriptomes per cell type (level 3 data) will be used to determine whether the cognate receptor for each ligand in the previous experiment is present.  Ligand-receptor interactions are from the Bader lab database included in [CCInx](https://baderlab.github.io/CCInx/).

```{r load_data_lvl3_ctl,echo=FALSE,message=FALSE,warning=FALSE,include=TRUE}
if (exists("lvl3_ctl")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData") 
} else {
  source("lvl3_inputs_allgenesctrl.R")
}

temp <- load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_human.RData",package="CCInx"))

lig16R <- sapply(lig16,function(X) {
  temp <- c(inxDB$nodeB[inxDB$nodeA == X],inxDB$nodeA[inxDB$nodeB == X])
  return(temp[grepl("Receptor",geneInfo[temp,"protein_type"])])
},simplify=F)
lig16R[["GAS6"]] <- c("AXL","TYRO3","MERTK")
for (LIG in lig16) {
  temp_in <- lig16R[[LIG]] %in% lvl3_ctl@rdesc$pr_gene_symbol
  lig16R[[LIG]] <- lig16R[[LIG]][temp_in]
  names(lig16R[[LIG]]) <- sapply(lig16R[[LIG]],function(X) 
    lvl3_ctl@rdesc$id[lvl3_ctl@rdesc$pr_gene_symbol == X])
  cat(paste0(round(sum(temp_in) / length(temp_in) * 100),"% (",
               round(sum(lvl3_ctl@rdesc[names(lig16R[[LIG]]),"pr_is_lm"] == "1") / length(temp_in) * 100),
               "%) of ",LIG," receptors present in CMap data, either inferred or (assayed)."),"\n")
}

rm(list=c("LIG","temp",temp))
```

Below is an overview of all potential cognate receptors per ligand present in the data.  The left figure shows which ligands (dark purple) each receptor is known to interact with.  On the right is a heatmap showing mean normalized gene expression per cell type for each receptor gene, with darker indicating higher expression.  Gene names (rows) are ordered by heirarchical clustering of ligand-receptor interaction data, such that receptors common to the same ligand should be together.  **Landmark** gene expression was emperically determined rather than inferred (names indicated in bold).  Ligand names (columns, left) are ordered by heirarchical clustering of ligand-receptor interaction data, such that ligands sharing the same receptors should be proximal.  Cell type names (columns, right) are ordered by heirarchical clustering of mean receptor gene expression.

```{r rec_expr,echo=F,message=F,warning=F,fig.height=25,fig.width=9}
temp_genes <- unique(unlist(lapply(lig16R,names)))
names(temp_genes) <- lvl3_ctl@rdesc[temp_genes,"pr_gene_symbol"]
lig16Rmat <- sapply(lig16R,function(X) names(temp_genes) %in% X)
rownames(lig16Rmat) <- names(temp_genes)
temp_roworder <- hclust(dist(lig16Rmat))$order
lig16Rmat <- lig16Rmat[temp_roworder,]
lig16Rmat <- lig16Rmat[,hclust(dist(t(lig16Rmat)))$order]

meanR <- sapply(ct14,function(CT) rowMeans(lvl3_ctl@mat[temp_genes,lvl3_ctl@cdesc$cell_id == CT]))
rownames(meanR) <- names(temp_genes)
meanR <- meanR[temp_roworder,]
meanR <- meanR[,hclust(dist(t(meanR)))$order]

layout(cbind(rep(1,25),rep(2,25),c(rep(3,5),rep(0,20))),widths=c(3,3,1))
par(mar=c(6,0.5,6,5.5),mgp=2:0,las=2)
image(z=t(lig16Rmat[rev(rownames(lig16Rmat)),]),
      x=1:ncol(lig16Rmat),y=1:nrow(lig16Rmat),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(10,palette="inferno",rev=T)[c(1,9)])
mtext(colnames(lig16Rmat),side=3,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
mtext(colnames(lig16Rmat),side=1,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
title("Ligand interaction with receptors",line=4.5)

par(mar=c(6,0,6,0.5),mgp=2:0,las=2)
image(z=t(meanR[rev(rownames(meanR)),]),
      x=1:ncol(meanR),y=1:nrow(meanR),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(100,palette="inferno",rev=T))
mtext(ct14[colnames(meanR)],side=3,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(ct14[colnames(meanR)],side=1,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(rev(rownames(meanR)),xpd=NA,side=2,line=0.2,at=1:nrow(meanR),cex=0.7,
      font=rev(as.integer(lvl3_ctl@rdesc[temp_genes,"pr_is_lm"]) + 1))

par(mar=c(0,3.5,6,0.5),mgp=2:0,las=0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",axes=F,
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",side=1,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```

```{r rec_per_lig, echo=F,message=F,warning=F,fig.height=6,fig.width=9}
maxRL <- sapply(lig16R,function(X) apply(meanR[X,],2,max))
meanRL <- sapply(lig16R,function(X) apply(meanR[X,],2,mean))


maxRL <- maxRL[hclust(dist(maxRL),method="ward.D2")$order,
               hclust(dist(t(maxRL)),method="ward.D2")$order]
lig16Rmat <- lig16Rmat[,colnames(maxRL)]

layout(rbind(1:3),widths=c(1,3,2))
par(mar=c(9,3.5,2,1),mgp=2:0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",line=2.5,side=1)


par(mar=c(9,1,2,0))
image(z=maxRL,
      x=1:nrow(maxRL),y=1:ncol(maxRL),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      breaks=seq(0,15,length.out=1001),
      col=sequential_hcl(1000,palette="inferno",rev=T))
mtext("Max cognate receptor expression",font=2,line=0.5)
text(x=1:nrow(maxRL),y=rep(0,nrow(maxRL)),
     labels=rownames(maxRL),xpd=NA,srt=315,adj=0)

par(mar=c(9,5.5,2,1))
temp <- barplot(colSums(lig16Rmat),yaxs="i",axisnames=F,horiz=T)
mtext("Cognate receptors",line=2.5,side=1)
mtext(colnames(lig16Rmat),at=temp,
      side=2,line=2.5,adj=0.5,las=2)


```


## Receptor availability affecting ligand prediction accuracy

```{r load_data_lvl5, message=FALSE, warning=FALSE, include=FALSE}
if (exists("lvl5_data")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl5_inputs.RData") 
} else {
  source("lvl5_inputs.R")
}
```

```{r RF_leave1out,echo=FALSE,message=FALSE,warning=FALSE}
if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/200524_lvl5_mixall_balanced_leave1out_results.RData")) {
  temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/200524_lvl5_mixall_balanced_leave1out_results.RData")
} else {
  source("200524_lvl5_mixall_leave1out.R")
}

test_results <- data.frame(
  IDs=as.vector(testIDs),
  Labels=as.vector(t(sapply(rfresults,function(X) X$predictions))),
  stringsAsFactors=F
)
test_hits <- vector("list",sum(lvl5_data@cdesc$pert_iname %in% lig16))
names(test_hits) <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname %in% lig16]

for (X in 1:nrow(test_results)) {
  test_hits[[test_results[X,"IDs"]]] <- append(test_hits[[test_results[X,"IDs"]]],
                                               test_results[X,"Labels"])
}

acc_leave1out <- sapply(names(test_hits),function(X) 
  sum(test_hits[[X]] == lvl5_data@cdesc[X,"pert_iname"]) / 
    length(test_hits[[X]]))
acc_leave1out_totals <- sapply(lig16,function(X) {
  Y <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X]
  return(sum(unlist(test_hits[Y]) == X) / 
           length(unlist(test_hits[Y])))
})

acc_leave1out_ct <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    Z <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X &
                                     lvl5_data@cdesc$cell_id == Y]
    return(sum(unlist(test_hits[Z]) == X) / 
             length(unlist(test_hits[Z])))
  })
})

test_counts <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    sum(lvl5_data@cdesc$pert_iname == X &
          lvl5_data@cdesc$cell_id == Y)
  })
})

test_predictions <- sapply(lig16,function(X) {
  sapply(ct14,function(Y) {
    temp <- rownames(lvl5_data@cdesc)[lvl5_data@cdesc$pert_iname == X &
                                        lvl5_data@cdesc$cell_id == Y]
    return(table(factor(unlist(test_hits[temp]),levels=lig16)) / 
             length(unlist(test_hits[temp])))
  })
},simplify=F)

rm(X,temp)
```

Cognate receptor expression for each ligand was compared to the accuracy per cell line of the leave-one-out ligand classification model outlined [previously](https://baderlab.github.io/Brendan_CCInxPred/LigPred_lvl5.html#cell-type-biases-in-accuracy).  Lines for each receptor are coloured by the spearman correlation coefficient between expression and accuracy per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.

```{r rec_expr_acc_corr,echo=FALSE,message=FALSE,warning=FALSE,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,xlim=range(acc_leave1out_ct[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
       xlab="Prediction accuracy per cell type",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  segments(x0=seq(0,1,length.out=1000),
           x1=seq(0,1,length.out=1000),
           y0=rep(par("usr")[3],1000),
           y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
           lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))
  
  temp_order <- rownames(acc_leave1out_ct)[order(acc_leave1out_ct[,LIG])]
  temp_corr <- apply(meanR[lig16R[[LIG]],temp_order],1,function(X)
    cor(acc_leave1out_ct[temp_order,LIG],X,method="spearman"))
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(temp_corr)]) {
    points(acc_leave1out_ct[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(temp_corr)[order(temp_corr,decreasing=T)][1:3]
  text(acc_leave1out_ct[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(acc_leave1out_ct[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r temp_IFNG, eval=FALSE,echo=FALSE,message=FALSE,warning=FALSE}
LIG <- "IFNG"
par(mar=c(4,6.5,1,3.5),mgp=2:0)
plot(NA,NA,xlim=range(acc_leave1out_ct[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
     xlab="Prediction accuracy per cell type",ylab=NA)
axis(2,line=3.5)
mtext("Receptor expression per cell type",
      side=2,line=5.5,cex=0.7)
title(LIG,line=0)
segments(x0=seq(0,1,length.out=1000),
         x1=seq(0,1,length.out=1000),
         y0=rep(par("usr")[3],1000),
         y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
         lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))

temp_order <- rownames(acc_leave1out_ct)[order(acc_leave1out_ct[,LIG])]
temp_corr <- apply(meanR[lig16R[[LIG]],temp_order],1,function(X)
  cor(acc_leave1out_ct[temp_order,LIG],X,method="spearman"))
temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
names(temp_col) <- lig16R[[LIG]]
for (GENE in lig16R[[LIG]][order(temp_corr)]) {
  points(acc_leave1out_ct[temp_order,LIG],
         meanR[GENE,temp_order],
         type="b",pch=20,col=temp_col[GENE],lwd=2)
}
temp_top <- c("IFNGR1","IFNGR2")
for (GENE in temp_top) {
  points(acc_leave1out_ct[temp_order,LIG],
         meanR[GENE,temp_order],
         type="b",pch=20,col="red",lwd=2)
}

text(acc_leave1out_ct[temp_order[length(temp_order)],LIG],
     meanR[temp_top,temp_order[length(temp_order)]],
     labels=temp_top,col=temp_col[temp_top],
     pos=4,xpd=NA,offset=0.4,
     font=as.integer(sapply(temp_top,function(X) 
       lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
text(acc_leave1out_ct[temp_order[1],LIG],
     meanR[temp_top,temp_order[1]],
     labels=temp_top,col=temp_col[temp_top],
     pos=2,xpd=NA,offset=0.4,
     font=as.integer(sapply(temp_top,function(X) 
       lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)


```

```{r temp_EGF, eval=FALSE,echo=FALSE,message=FALSE,warning=FALSE}
LIG <- "EGF"

# grep("EGFR",lvl3_ctl@rdesc$pr_gene_symbol,value=TRUE)

par(mar=c(4,6.5,1,3.5),mgp=2:0)
plot(NA,NA,xlim=range(acc_leave1out_ct[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
     xlab="Prediction accuracy per cell type",ylab=NA)
axis(2,line=3.5)
mtext("Receptor expression per cell type",
      side=2,line=5.5,cex=0.7)
title(LIG,line=0)
segments(x0=seq(0,1,length.out=1000),
         x1=seq(0,1,length.out=1000),
         y0=rep(par("usr")[3],1000),
         y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
         lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))

temp_order <- rownames(acc_leave1out_ct)[order(acc_leave1out_ct[,LIG])]
temp_corr <- apply(meanR[lig16R[[LIG]],temp_order],1,function(X)
  cor(acc_leave1out_ct[temp_order,LIG],X,method="spearman"))
temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,temp_corr),100,labels=F)[c(-1,-2)]]
names(temp_col) <- lig16R[[LIG]]
for (GENE in lig16R[[LIG]][order(temp_corr)]) {
  points(acc_leave1out_ct[temp_order,LIG],
         meanR[GENE,temp_order],
         type="b",pch=20,col=temp_col[GENE],lwd=2)
}
temp_top <- c("EGFR")
for (GENE in temp_top) {
  points(acc_leave1out_ct[temp_order,LIG],
         meanR[GENE,temp_order],
         type="b",pch=20,col="red",lwd=2)
}

text(acc_leave1out_ct[temp_order[length(temp_order)],LIG],
     meanR[temp_top,temp_order[length(temp_order)]],
     labels=temp_top,col=temp_col[temp_top],
     pos=4,xpd=NA,offset=0.4,
     font=as.integer(sapply(temp_top,function(X) 
       lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
text(acc_leave1out_ct[temp_order[1],LIG],
     meanR[temp_top,temp_order[1]],
     labels=temp_top,col=temp_col[temp_top],
     pos=2,xpd=NA,offset=0.4,
     font=as.integer(sapply(temp_top,function(X) 
       lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)


```



****

### Tumour necrosis factor alpha (TNF)

TNF binds to two receptors, TNFR1 and TNFR2, neither of which are present in the CMap data (neither assayed nor inferred).  This is probably because they are (at least TNFR1) ubiquitously expressed, and thus their expression profiles are not informative or easily inferred.  That may also explain why the transcriptional response to TNF treatment is so consistent - the may be receptor is present on all cell types.  Downstream of TNF receptor activation, the multiple signaling pathways with conflicting phenotypes can be activated, though our data shows robust increases in expression of a common subset of genes across cell lines (see above).

****