---
title: "Biological factors affecting prediction accuracy"
---
  
```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
library(cmapR)
library(ranger)
library(colorspace)
library(kableExtra)
source("~/Dropbox/GDB/line2user.R")
knitr::opts_chunk$set(
  echo=F,
  message=F,
  warning=F
)
```

# 16 ligands

## Receptor gene expression

The quantile-normalized transcriptomes per cell type (level 3 data) will be used to determine whether the cognate receptor for each ligand in the previous experiment is present.  Ligand-receptor interactions are from the Bader lab database included in [CCInx](https://baderlab.github.io/CCInx/).

```{r load_data_lvl3_ctl,include=TRUE}
if (exists("lvl3_ctl")) {
} else if (file.exists("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData")) {
  load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData") 
} else {
  source("lvl3_inputs_allgenesctrl.R")
}

temp <- load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_human.RData",package="CCInx"))

lig16R <- sapply(lig16,function(X) {
  temp <- c(inxDB$nodeB[inxDB$nodeA == X],inxDB$nodeA[inxDB$nodeB == X])
  return(temp[grepl("Receptor",geneInfo[temp,"protein_type"])])
},simplify=F)
lig16R[["GAS6"]] <- c("AXL","TYRO3","MERTK")
for (LIG in lig16) {
  temp_in <- lig16R[[LIG]] %in% lvl3_ctl@rdesc$pr_gene_symbol
  lig16R[[LIG]] <- lig16R[[LIG]][temp_in]
  names(lig16R[[LIG]]) <- sapply(lig16R[[LIG]],function(X) 
    lvl3_ctl@rdesc$id[lvl3_ctl@rdesc$pr_gene_symbol == X])
  cat(paste0(round(sum(temp_in) / length(temp_in) * 100),"% (",
               round(sum(lvl3_ctl@rdesc[names(lig16R[[LIG]]),"pr_is_lm"] == "1") / length(temp_in) * 100),
               "%) of ",LIG," receptors present in CMap data, either inferred or (assayed)."),"\n")
}

rm(list=c("LIG","temp",temp))
```

Below is an overview of all potential cognate receptors per ligand present in the data.  The left figure shows which ligands (dark purple) each receptor is known to interact with.  On the right is a heatmap showing mean normalized gene expression per cell type for each receptor gene, with darker indicating higher expression.  Gene names (rows) are ordered by heirarchical clustering of ligand-receptor interaction data, such that receptors common to the same ligand should be together.  **Landmark** gene expression was emperically determined rather than inferred (names indicated in bold).  Ligand names (columns, left) are ordered by heirarchical clustering of ligand-receptor interaction data, such that ligands sharing the same receptors should be proximal.  Cell type names (columns, right) are ordered by heirarchical clustering of mean receptor gene expression.

```{r rec_expr,echo=F,message=F,warning=F,fig.height=25,fig.width=9}
temp_genes <- unique(unlist(lapply(lig16R,names)))
names(temp_genes) <- lvl3_ctl@rdesc[temp_genes,"pr_gene_symbol"]
lig16Rmat <- sapply(lig16R,function(X) names(temp_genes) %in% X)
rownames(lig16Rmat) <- names(temp_genes)
temp_roworder <- hclust(dist(lig16Rmat))$order
lig16Rmat <- lig16Rmat[temp_roworder,]
lig16Rmat <- lig16Rmat[,hclust(dist(t(lig16Rmat)))$order]

meanR <- sapply(ct14,function(CT) rowMeans(lvl3_ctl@mat[temp_genes,lvl3_ctl@cdesc$cell_id == CT]))
rownames(meanR) <- names(temp_genes)
meanR <- meanR[temp_roworder,]
meanR <- meanR[,hclust(dist(t(meanR)))$order]

layout(cbind(rep(1,25),rep(2,25),c(rep(3,5),rep(0,20))),widths=c(3,3,1))
par(mar=c(6,0.5,6,5.5),mgp=2:0,las=2)
image(z=t(lig16Rmat[rev(rownames(lig16Rmat)),]),
      x=1:ncol(lig16Rmat),y=1:nrow(lig16Rmat),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(10,palette="inferno",rev=T)[c(1,9)])
mtext(colnames(lig16Rmat),side=3,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
mtext(colnames(lig16Rmat),side=1,line=0.2,at=1:ncol(lig16Rmat),cex=0.8)
title("Ligand interaction with receptors",line=4.5)

par(mar=c(6,0,6,0.5),mgp=2:0,las=2)
image(z=t(meanR[rev(rownames(meanR)),]),
      x=1:ncol(meanR),y=1:nrow(meanR),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      col=sequential_hcl(100,palette="inferno",rev=T))
mtext(ct14[colnames(meanR)],side=3,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(ct14[colnames(meanR)],side=1,line=0.2,at=1:ncol(meanR),cex=0.8)
mtext(rev(rownames(meanR)),xpd=NA,side=2,line=0.2,at=1:nrow(meanR),cex=0.7,
      font=rev(as.integer(lvl3_ctl@rdesc[temp_genes,"pr_is_lm"]) + 1))

par(mar=c(0,3.5,6,0.5),mgp=2:0,las=0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",axes=F,
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",side=1,cex=0.7)

rm(list=grep("^temp",ls(),value=T))
```

```{r rec_per_lig,fig.height=6,fig.width=9}
maxRL <- sapply(lig16R,function(X) apply(meanR[X,],2,max))
meanRL <- sapply(lig16R,function(X) apply(meanR[X,],2,mean))


maxRL <- maxRL[hclust(dist(maxRL),method="ward.D2")$order,
               hclust(dist(t(maxRL)),method="ward.D2")$order]
lig16Rmat <- lig16Rmat[,colnames(maxRL)]

layout(rbind(1:3),widths=c(1,3,2))
par(mar=c(9,3.5,2,1),mgp=2:0)
temp_hist <- hist(meanR,breaks=seq(0,15,length.out=101),plot=F)
barplot(temp_hist$counts,horiz=T,yaxs="i",
        col=sequential_hcl(100,palette="inferno",rev=T),
        border=sequential_hcl(100,palette="inferno",rev=T))
axis(side=2,at=seq(par("usr")[3],par("usr")[4],length.out=4),
     labels=seq(0,15,length.out=4))
mtext("Mean normalized receptor expression per cell type",side=2,font=2,line=2)
mtext("Frequency",line=2.5,side=1)


par(mar=c(9,1,2,0))
image(z=maxRL,
      x=1:nrow(maxRL),y=1:ncol(maxRL),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA,
      breaks=seq(0,15,length.out=1001),
      col=sequential_hcl(1000,palette="inferno",rev=T))
mtext("Max cognate receptor expression",font=2,line=0.5)
text(x=1:nrow(maxRL),y=rep(0,nrow(maxRL)),
     labels=rownames(maxRL),xpd=NA,srt=315,adj=0)

par(mar=c(9,5.5,2,1))
temp <- barplot(colSums(lig16Rmat),yaxs="i",axisnames=F,horiz=T)
mtext("Cognate receptors",line=2.5,side=1)
mtext(colnames(lig16Rmat),at=temp,
      side=2,line=2.5,adj=0.5,las=2)


```
```{r lig_DE,fig.height=6,fig.width=9}
temp <- load("~/Dropbox/GDB_archive/CMapCorr_files/lig16_DE_ligct_FDR.RData")

scoreDE <- pDE_ligct[,"01"] + 1e-4
scoreDE[scoreDE > 1] <- 1
scoreDE <- -log10(scoreDE)
  
scoreMAT <- tapply(scoreDE,as.factor(sub("^.+_","",names(scoreDE))),c)
for (X in names(scoreMAT)) {
  names(scoreMAT[[X]]) <- sub(paste0("_",X),"",names(scoreMAT[[X]]))
  scoreMAT[[X]] <- scoreMAT[[X]][sort(names(scoreMAT[[X]]))]
}
scoreMAT <- t(do.call(rbind,scoreMAT))
scoreMAT <- scoreMAT[rownames(maxRL),colnames(maxRL)]


layout(matrix(c(1,1,2,3),nrow=2),widths=c(2,1))

par(mar=c(3,2,7,6),mgp=2:0,las=2)
image(scoreMAT,
      col=sequential_hcl(100,palette="PinkYl",rev=T),
      x=1:nrow(scoreMAT),y=1:ncol(scoreMAT),
      xaxt="n",yaxt="n",xlab=NA,ylab=NA)
mtext(colnames(scoreMAT),side=4,line=0.1,
      at=1:ncol(scoreMAT),adj=0,cex=switch((ncol(scoreMAT) > 20) + 1,0.9,0.5),
      col=qualitative_hcl(ncol(scoreMAT),palette="dark3"))
mtext("Ligands",side=4,line=4.5,las=0,font=2,cex=1.5,
      at=par("usr")[3] + (par("usr")[4] - par("usr")[3]) / 2,adj=0.5)
text(x=1:nrow(scoreMAT),y=rep(line2user(0.3,3),nrow(scoreMAT)),
     col=qualitative_hcl(nrow(scoreMAT),palette="dark3"),
     labels=rownames(scoreMAT),xpd=NA,adj=0,srt=45,cex=0.8)
text(x=0,y=line2user(1,3),labels="Cell lines",
     xpd=NA,adj=0,srt=45,font=2,cex=1.5)

rect(xleft=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,1] - 0.5,
     xright=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,1] + 0.5,
     ybottom=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,2] - 0.5,
     ytop=which(scoreMAT >= -log10(0.05 + 1e-4),arr.ind=T)[,2] + 0.5,
     border="dodgerblue")

temp_left <- par("usr")[1] + (par("usr")[2] - par("usr")[1]) * 0.3
temp_right <- par("usr")[1] + (par("usr")[2] - par("usr")[1]) * 0.7

segments(x0=seq(temp_left,temp_right,length.out=1000),
         x1=seq(temp_left,temp_right,length.out=1000),
         y0=rep(line2user(1,1),1000),y1=rep(line2user(1.6,1),1000),
         xpd=NA,col=sequential_hcl(1000,palette="PinkYl",rev=T))
mtext(c("100%","< 0.01%"),side=1,line=0.8,las=0,adj=c(1.1,-0.1),
      at=c(temp_left,temp_right))
mtext("Probability of at least # DE occuring by chance",
      las=0,side=1,line=1.7,adj=0.5,
      at=temp_left + (temp_right - temp_left) / 2)

rect(xleft=temp_left + (-log10(0.05 + 1e-4) / max(scoreMAT)) * (temp_right - temp_left),
     xright=temp_right,ybottom=line2user(1,1),ytop=line2user(1.6,1),
     xpd=NA,col=NA,border="dodgerblue")
mtext("p <= 0.05",side=1,las=0,col="dodgerblue",cex=0.9,line=0,adj=1,at=temp_right)

rm(list=temp)
rm(list=c("X",grep("^temp",ls(),value=T)))



par(mar=c(3,3,2,1),mgp=2:0,las=0)
plot(as.vector(maxRL),as.vector(scoreMAT),pch=20,
     col=rep(qualitative_hcl(nrow(maxRL),palette="dark3"),ncol(maxRL)), # per cell line
     xlab="Max cognate receptor expression",
     ylab="# DE probability score")
mtext("Coloured by cell line",font=2,line=0.1)
mtext(paste("PCC =",signif(cor(as.vector(maxRL),as.vector(scoreMAT)),2)),
      line=-1,adj=0.05,cex=0.9)
plot(as.vector(maxRL),as.vector(scoreMAT),pch=20,
     col=as.vector(sapply(qualitative_hcl(ncol(maxRL),palette="dark3"),rep,times=nrow(maxRL))), # per ligand
     xlab="Max cognate receptor expression",
     ylab="# DE probability score")
mtext("Coloured by ligand",font=2,line=0.1)
mtext(paste("PCC =",signif(cor(as.vector(maxRL),as.vector(scoreMAT)),2)),
      line=-1,adj=0.05,cex=0.9)

```

## Receptor availability affecting transcriptional response to ligand treatment

Cognate receptor expression for each ligand was compared to probability scores for the number of differentially expressed genes associated with each ligand-treated cell line outlined [previously](https://baderlab.github.io/Brendan_CCInxPred/DEoverlap_FDR.html#averaging-within-each-ligand-per-cell-line).  Lines for each receptor are coloured by the spearman correlation coefficient between expression and DE score per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.

```{r rec_DE_corr}
Rcor_scoreDE <- sapply(lig16,function(LIG)
  apply(meanR[lig16R[[LIG]],rownames(scoreMAT)],1,function(X) 
    cor(X,scoreMAT[,LIG],method="spearman")))

```



```{r rec_DE_corr_detail,fig.height=9,fig.width=9,fig.show='hold'}
par(mfrow=c(4,4),mar=c(4,6.5,1,3.5),mgp=2:0)
for (LIG in lig16) {
  plot(NA,NA,xlim=range(scoreMAT[,LIG]),ylim=c(0,15),bty="n",yaxt="n",
       xlab="# DE probability score",ylab=NA)
  axis(2,line=3.5)
  mtext("Receptor expression per cell type",
        side=2,line=5.5,cex=0.7)
  title(LIG,line=0)
  segments(x0=seq(min(scoreMAT),max(scoreMAT),length.out=1000),
           x1=seq(min(scoreMAT),max(scoreMAT),length.out=1000),
           y0=rep(par("usr")[3],1000),
           y1=rep(par("usr")[3] + (par("usr")[4] - par("usr")[3]) * 0.02,1000),
           lwd=2,col=sequential_hcl(1000,palette="inferno",rev=T))

  temp_order <- rownames(scoreMAT)[order(scoreMAT[,LIG])]
  temp_col <- diverge_hcl(100,palette="Purple-Green")[cut(c(-1,1,Rcor_scoreDE[[LIG]]),100,labels=F)[c(-1,-2)]]
  names(temp_col) <- lig16R[[LIG]]
  for (GENE in lig16R[[LIG]][order(Rcor_scoreDE[[LIG]])]) {
    points(scoreMAT[temp_order,LIG],
           meanR[GENE,temp_order],
           type="b",pch=20,col=temp_col[GENE],lwd=2)
  }
  temp_top <- names(Rcor_scoreDE[[LIG]])[order(Rcor_scoreDE[[LIG]],decreasing=T)][1:3]
  text(scoreMAT[temp_order[length(temp_order)],LIG],
       meanR[temp_top,temp_order[length(temp_order)]],
       labels=temp_top,col=temp_col[temp_top],
       pos=4,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  text(scoreMAT[temp_order[1],LIG],
       meanR[temp_top,temp_order[1]],
       labels=temp_top,col=temp_col[temp_top],
       pos=2,xpd=NA,offset=0.4,
       font=as.integer(sapply(temp_top,function(X) 
         lvl3_ctl@rdesc[lvl3_ctl@rdesc$pr_gene_symbol == X,"pr_is_lm"])) + 1)
  
}
rm(list=c("LIG","GENE",grep("^temp",ls(),value=T)))
```

```{r}
temp_lig <- qualitative_hcl(length(lig16),palette="dark3")[as.vector(rbind(1:8,9:16))]
names(temp_lig) <- lig16
plot(NA,NA,xlim=range(scoreMAT),ylim=c(0,15),
     xlab="# DE probability score",
     ylab="Receptor expression per cell type")
for (LIG in lig16) {
  CT <- rownames(scoreMAT)[order(scoreMAT[,LIG])]
  points(scoreMAT[CT,LIG],
         meanR[names(which.max(Rcor_scoreDE[[LIG]])),CT],
         type="b",pch=20,col=temp_lig[LIG])
}
```

## Linear model

```{r load_data_lvl3}
rm(list=ls()[!ls() %in% c("line2user","lig16R","ct14","lig16")])
load("~/Dropbox/GDB_archive/CMapCorr_files/lig16_DE_ligct_FDR.RData")
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_inputs_allgenes.RData")
if (identical(rownames(lvl3_data_all@mat),rownames(lvl3_data_all_ctl@mat))) {
  lvl3_data <- GCT(mat=cbind(lvl3_data_all_ctl@mat,lvl3_data_all@mat),
                   rdesc=lvl3_data_all_ctl@rdesc,
                   cdesc=rbind(lvl3_data_all_ctl@cdesc,lvl3_data_all@cdesc),
                   rid=lvl3_data_all_ctl@rid,
                   cid=c(lvl3_data_all_ctl@cid,lvl3_data_all@cid))
  rm(lvl3_data_all,lvl3_data_all_ctl)
} else {
  stop("something's fucky")
}
```

```{r}
LIG <- "EGF"
CT <- c("VCAP","HCC515")
X <- names(lig16R[[LIG]][1])
Y <- "1958"

plot(lvl3_data@mat[X,],lvl3_data@mat[Y,],pch=".")
points(lvl3_data@mat[X,lvl3_data@cdesc$pert_iname == LIG],
       lvl3_data@mat[Y,lvl3_data@cdesc$pert_iname == LIG],
       pch=".",cex=2,col="red")
points(lvl3_data@mat[X,lvl3_data@cdesc$pert_iname == LIG & lvl3_data@cdesc$cell_id %in% CT],
       lvl3_data@mat[Y,lvl3_data@cdesc$pert_iname == LIG & lvl3_data@cdesc$cell_id %in% CT],
       pch=".",cex=3,col="dodgerblue")

```


# 295 ligands 

```{r load_data_lig295}
rm(list=ls())
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl4_inputs_allgenes.RData")
lvl4_data <- lvl4_data_all
rm(lvl4_data_all)

temp_ligcountsperct <- sapply(unique(lvl4_data@cdesc$cell_id),function(CT)
  length(unique(lvl4_data@cdesc[lvl4_data@cdesc$cell_id == CT,"pert_iname"])))
ct9 <- names(temp_ligcountsperct)[temp_ligcountsperct > 100]
names(ct9) <- sapply(ct9,function(X) names(ct14)[ct14 == X])

temp_ctcountsperlig <- sapply(unique(lvl4_data@cdesc$pert_iname),function(LIG)
  length(unique(lvl4_data@cdesc[lvl4_data@cdesc$pert_iname == LIG,"cell_id"])))

lig295 <- names(temp_ctcountsperlig)[temp_ctcountsperlig >= 9]

rm(list=ls()[!ls() %in% c("ct9","lig295")])
source("~/Dropbox/GDB/line2user.R")
```

```{r load_data_recp}
load("~/Dropbox/GDB_archive/CMapCorr_files/lvl3_allgenesctrl.RData")
temp <- load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_human.RData",package="CCInx"))

lig295 <- lig295[lig295 %in% geneInfo$hgnc_symbol]

lig295R <- sapply(lig295,function(X) {
  temp <- c(inxDB$nodeB[inxDB$nodeA == X],inxDB$nodeA[inxDB$nodeB == X])
  return(temp[grepl("Receptor",geneInfo[temp,"protein_type"])])
},simplify=F)
for (LIG in lig295) {
  temp_in <- lig295R[[LIG]] %in% lvl3_ctl@rdesc$pr_gene_symbol
  lig295R[[LIG]] <- lig295R[[LIG]][temp_in]
  names(lig295R[[LIG]]) <- sapply(lig295R[[LIG]],function(X) 
    lvl3_ctl@rdesc$id[lvl3_ctl@rdesc$pr_gene_symbol == X])
}
lig295R <- lig295R[sapply(lig295R,length) > 0]
lig295 <- names(lig295R)

rm(list=c(temp,"lig16","ct14","LIG","temp_in","temp"))
```

```{r}

```

