<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Re-analysis of NicheNet dataset</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CCinxPred?</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Old Analysis Attempts
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="CmapDataSummary.html">CMap Data</a>
    </li>
    <li>
      <a href="CmapCtrlCorr.html">Sanity Check</a>
    </li>
    <li>
      <a href="CellStimPred_lvl3.html">Tx pred</a>
    </li>
    <li>
      <a href="LigPred_lvl5.html">Lvl5 lig pred</a>
    </li>
    <li>
      <a href="RecExpr.html">Lvl5 followup</a>
    </li>
    <li>
      <a href="LigPred_lvl4.html">Lvl4 lig pred</a>
    </li>
    <li>
      <a href="LigPred_extrapolate.html">Lig pred on new CT</a>
    </li>
    <li>
      <a href="LigProb_NovelLines.html">Lig prob on new CT</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Figures For Paper
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Fig1.html">Fig 1</a>
    </li>
    <li>
      <a href="Fig1_allgenes.html">Fig 1 (all genes)</a>
    </li>
    <li>
      <a href="Fig2.html">Fig 2</a>
    </li>
    <li>
      <a href="Fig2_allgenes.html">Fig 2 (all genes)</a>
    </li>
    <li>
      <a href="Fig3.html">Fig 3</a>
    </li>
    <li>
      <a href="Fig3_allgenes.html">Fig 3 (all genes)</a>
    </li>
    <li>
      <a href="Fig4.html">Fig 4</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Paper Figs v2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Zcorr.html">Z-score correlation</a>
    </li>
    <li>
      <a href="GSEAcorr.html">GSEA results</a>
    </li>
    <li>
      <a href="DEoverlap.html">DE overlap</a>
    </li>
    <li>
      <a href="DEoverlap_FDR.html">DE overlap (FDR)</a>
    </li>
    <li>
      <a href="RecExpr_FDR.html">Receptor availability</a>
    </li>
    <li>
      <a href="NicheNet.html">NicheNet</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Paper Figs v3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="DEoverlap_FDR_forPaper.html">Fig 1 &amp; 2: DE overlap (FDR)</a>
    </li>
    <li>
      <a href="Zcorr_forPaper.html">Fig 3: GSEA (&amp; Z-score) correlation</a>
    </li>
    <li>
      <a href="RFresults_forPaper.html">Fig 4: RF results</a>
    </li>
    <li>
      <a href="NicheNet_forPaper.html">Fig 5: NicheNet</a>
    </li>
    <li>
      <a href="GSEAcorr2.html">Final GSEA stuff</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Re-analysis of NicheNet dataset</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#nichenet-calculation-sanity-check">NicheNet calculation sanity check</a></li>
<li><a href="#nichenet-data-summary">NicheNet data summary</a></li>
<li><a href="#comparing-across-vs-within-cell-type-per-ligand-treatment">Comparing across vs within cell type per ligand treatment</a><ul>
<li><a href="#correlation">Correlation</a></li>
<li><a href="#differential-gene-expression">Differential gene expression</a></li>
</ul></li>
<li><a href="#using-all-data-including-time-series">Using all data (including time-series)</a><ul>
<li><a href="#correlation-1">Correlation</a></li>
<li><a href="#differential-gene-expression-1">Differential gene expression</a></li>
</ul></li>
<li><a href="#after-removing-time-series-data">After removing time-series data</a><ul>
<li><a href="#correlation-2">Correlation</a></li>
<li><a href="#differential-gene-expression-2">Differential gene expression</a></li>
</ul></li>
</ul>
</div>

<div id="nichenet-calculation-sanity-check" class="section level1">
<h1>NicheNet calculation sanity check</h1>
<p><img src="NicheNet_files/figure-html/plot_raw-1.png" width="576" /></p>
<p>Looks like the sign of the logFC is flipped in the NicheNet calculations. Multiplying logFC values from the NicheNet dataset by -1 for the remainder of this analysis.</p>
<hr />
</div>
<div id="nichenet-data-summary" class="section level1">
<h1>NicheNet data summary</h1>
<p><img src="NicheNet_files/figure-html/DSsummary-1.png" width="672" /></p>
<p>18 ligands were assayed in multiple datasets, 11 of which occurred across more than one cell type.</p>
<hr />
</div>
<div id="comparing-across-vs-within-cell-type-per-ligand-treatment" class="section level1">
<h1>Comparing across vs within cell type per ligand treatment</h1>
<p>Previous analysis (see below) included a “within lab” category that was unclear, so it has been removed.<br />
All data (including time-series) is included in this analysis. NicheNet calculation of time-series data seems to result in all positive logFC values, so that makes some things a little screwy. In those cases, time-series data was only compared to other time-series data.</p>
<div id="correlation" class="section level2">
<h2>Correlation</h2>
<p><img src="NicheNet_files/figure-html/AvC_corr_plot-1.png" width="768" /></p>
</div>
<div id="differential-gene-expression" class="section level2">
<h2>Differential gene expression</h2>
<p>Differentially expressed genes were determined by |logFC| &gt;= 1 and FDR &lt;= 0.1 (similar to Browaeys <em>et al.</em>, 2020). This figure shows number of DE genes overlapping across all datasets testing the same ligand vs. datasets testing the same ligand in the same cell type vs. datasets from the same accession number (and same ligand / cell type). P-values for overlap (y-axis) were determined after determining probability of overlapping DE genes by chance through bootstrapping.</p>
<p><img src="NicheNet_files/figure-html/AvC_diffexp_boxplot_v2-1.png" width="864" /></p>
<p><img src="NicheNet_files/figure-html/AvC_diffexp_boxplot_v3-1.png" width="864" /></p>
<hr />
</div>
</div>
<div id="using-all-data-including-time-series" class="section level1">
<h1>Using all data (including time-series)</h1>
<p>NicheNet calculation of time-series data seems to result in all positive logFC values, so that makes some things a little screwy. In those cases, time-series data was only compared to other time-series data.</p>
<div id="correlation-1" class="section level2">
<h2>Correlation</h2>
<p><img src="NicheNet_files/figure-html/corr_boxplot-1.png" width="864" /></p>
<p><img src="NicheNet_files/figure-html/corr_boxplot_compTS-1.png" width="864" /> Despite the goofiness of the time-series logFC calculations (all are positive?), there’s no discernible difference in correlation of logFC within time-series experiments vs non-time-series.</p>
<p>Note that calculating correlation between time-series and non-TS logFC values is chaos, so those comparisons are skipped in the following figure.</p>
<div id="final-correlation-figure" class="section level3">
<h3>Final correlation figure</h3>
<p><img src="NicheNet_files/figure-html/corr_plot-1.png" width="864" /></p>
</div>
</div>
<div id="differential-gene-expression-1" class="section level2">
<h2>Differential gene expression</h2>
<div id="overlap-of-de-genes" class="section level3">
<h3>Overlap of DE genes</h3>
<p>Differentially expressed genes were determined by logFC &gt;= 1 and FDR &lt;= 0.1 (as in Browaeys <em>et al.</em>, 2020). This figure shows number of DE genes overlapping across all datasets testing the same ligand vs. datasets testing the same ligand in the same cell type vs. datasets from the same accession number (and same ligand / cell type). P-values for overlap (y-axis) were determined after determining probability of overlapping DE genes by chance through bootstrapping.</p>
<p><img src="NicheNet_files/figure-html/diffexp_boxplot-1.png" width="864" /></p>
<p><img src="NicheNet_files/figure-html/diffexp_boxplot_v2-1.png" width="864" /></p>
</div>
<div id="averaged-logfc-per-ligand" class="section level3">
<h3>Averaged logFC per ligand</h3>
<p>Similar to CMap analyses, took mean LogFC across all samples treated with the same ligand, or same ligand in the same cell line, or from the same accession number (and same ligand / cell line).</p>
<p><img src="NicheNet_files/figure-html/mean_LFC_hist-1.png" width="480" /></p>
</div>
<div id="averaged-logfc-per-ligand-compared-to-cmap" class="section level3">
<h3>Averaged logFC per ligand compared to CMap</h3>
<p>Connectivity Map data from the same set of ligands assayed in multiple experiments from the NicheNet datasets.</p>
<p><img src="NicheNet_files/figure-html/CMap_DE_load-1.png" width="864" /></p>
<p>Modified heatmaps showing logFC and FDR for selected genes from each sample from the NicheNet gold standard, as well as averages across all samples, all cell types, and all datasets where available. The figure below is an example for TNF.</p>
<p>
| <a href="output_figs/NN_BDNF.png">BDNF</a> | | <a href="output_figs/NN_BMP2.png">BMP2</a> | | <a href="output_figs/NN_CXCL12.png">CXCL12</a> | | <a href="output_figs/NN_EGF.png">EGF</a><a href="output_figs/NN_EGF_noTS.png">(no time-series)</a> |
</p>
<p>
| <a href="output_figs/NN_HGF.png">HGF</a> | | <a href="output_figs/NN_IFNA1.png">IFNA1</a> | | <a href="output_figs/NN_IFNG.png">IFNG</a><a href="output_figs/NN_IFNG_noTS.png">(no time-series)</a> | | <a href="output_figs/NN_IGF1.png">IGF1</a> |
</p>
<p>
| <a href="output_figs/NN_IL10.png">IL10</a> | | <a href="output_figs/NN_IL13.png">IL13</a><a href="output_figs/NN_IL13_noTS.png">(no time-series)</a> | | <a href="output_figs/NN_IL15.png">IL15</a> | | <a href="output_figs/NN_IL1B.png">IL1B</a><a href="output_figs/NN_IL1B_noTS.png">(no time-series)</a> |
</p>
<p>
| <a href="output_figs/NN_IL4.png">IL4</a><a href="output_figs/NN_IL4_noTS.png">(no time-series)</a> | | <a href="output_figs/NN_PDGFB.png">PDGFB</a> | | <a href="output_figs/NN_SHH.png">SHH</a><a href="output_figs/NN_SHH_noTS.png">(no time-series)</a> | | <a href="output_figs/NN_TGFB1.png">TGFB1</a><a href="output_figs/NN_TGFB1_noTS.png">(no time-series)</a> |
</p>
<p>
| <a href="output_figs/NN_TGFB3.png">TGFB3</a> | | <a href="output_figs/NN_TNF.png">TNF</a><a href="output_figs/NN_TNF_noTS.png">(no time-series)</a> | | <a href="output_figs/NN_VEGFA.png">VEGFA</a><a href="output_figs/NN_VEGFA_noTS.png">(no time-series)</a> | | <a href="output_figs/NN_WNT1.png">WNT1</a><a href="output_figs/NN_WNT1_noTS.png">(no time-series)</a> |
</p>
<p>
<img src="output_figs/NN_TNF_noTS.png"/>
</p>
</div>
<div id="comparing-nichenet-cmap-expression-changes-per-ligand" class="section level3">
<h3>Comparing NicheNet &amp; CMap expression changes per ligand</h3>
<p>Plots per ligand treatment of mean Z-score for each CMap cell line and logFC from NicheNet datasets.</p>
<p>
| <a href="output_figs/CMapNN_BMP2.png">BMP2</a> | | <a href="output_figs/CMapNN_EGF.png">EGF</a> |
</p>
<p>
| <a href="output_figs/CMapNN_IFNA1.png">IFNA1</a> | | <a href="output_figs/CMapNN_IFNG.png">IFNG</a> |
</p>
<p>
| <a href="output_figs/CMapNN_IL10.png">IL10</a> | | <a href="output_figs/CMapNN_IL13.png">IL13</a> | | <a href="output_figs/CMapNN_IL15.png">IL15</a> | | <a href="output_figs/CMapNN_IL1B.png">IL1B</a> |
</p>
<p>
| <a href="output_figs/CMapNN_IL4.png">IL4</a> | | <a href="output_figs/CMapNN_PDGFB.png">PDGFB</a> | | <a href="output_figs/CMapNN_TGFB1.png">TGFB1</a> |
</p>
<p>
| <a href="output_figs/CMapNN_TGFB3.png">TGFB3</a> | | <a href="output_figs/CMapNN_TNF.png">TNF</a> | | <a href="output_figs/CMapNN_VEGFA.png">VEGFA</a> | | <a href="output_figs/CMapNN_WNT1.png">WNT1</a> |
</p>
<p>
<img src="output_figs/CMapNN_IFNG.png"/>
</p>
<hr />
</div>
</div>
</div>
<div id="after-removing-time-series-data" class="section level1">
<h1>After removing time-series data</h1>
<div id="correlation-2" class="section level2">
<h2>Correlation</h2>
<p><img src="NicheNet_files/figure-html/noTS_corr_plot-1.png" width="864" /></p>
</div>
<div id="differential-gene-expression-2" class="section level2">
<h2>Differential gene expression</h2>
<p><img src="NicheNet_files/figure-html/noTS_diffexp_boxplot-1.png" width="864" /></p>
<p><img src="NicheNet_files/figure-html/noTS_diffexp_boxplot_v2-1.png" width="864" /></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
