<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Biological factors affecting transcriptional response to ligand stimulus</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">CCinxPred?</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Old Analysis Attempts
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="CmapDataSummary.html">CMap Data</a>
    </li>
    <li>
      <a href="CmapCtrlCorr.html">Sanity Check</a>
    </li>
    <li>
      <a href="CellStimPred_lvl3.html">Tx pred</a>
    </li>
    <li>
      <a href="LigPred_lvl5.html">Lvl5 lig pred</a>
    </li>
    <li>
      <a href="RecExpr.html">Lvl5 followup</a>
    </li>
    <li>
      <a href="LigPred_lvl4.html">Lvl4 lig pred</a>
    </li>
    <li>
      <a href="LigPred_extrapolate.html">Lig pred on new CT</a>
    </li>
    <li>
      <a href="LigProb_NovelLines.html">Lig prob on new CT</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Figures For Paper
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Fig1.html">Fig 1</a>
    </li>
    <li>
      <a href="Fig1_allgenes.html">Fig 1 (all genes)</a>
    </li>
    <li>
      <a href="Fig2.html">Fig 2</a>
    </li>
    <li>
      <a href="Fig2_allgenes.html">Fig 2 (all genes)</a>
    </li>
    <li>
      <a href="Fig3.html">Fig 3</a>
    </li>
    <li>
      <a href="Fig3_allgenes.html">Fig 3 (all genes)</a>
    </li>
    <li>
      <a href="Fig4.html">Fig 4</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Paper Figs v2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Zcorr.html">Z-score correlation</a>
    </li>
    <li>
      <a href="GSEAcorr.html">GSEA correlation</a>
    </li>
    <li>
      <a href="DEoverlap.html">DE overlap</a>
    </li>
    <li>
      <a href="DEoverlap_FDR.html">DE overlap (FDR)</a>
    </li>
    <li>
      <a href="RecExpr_FDR.html">Receptor availability</a>
    </li>
    <li>
      <a href="NicheNet.html">NicheNet</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Paper Figs v3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Zcorr_forPaper.html">Fig 1: Z-score correlation</a>
    </li>
    <li>
      <a href="DEoverlap_FDR_forPaper.html">Fig 2 &amp; 3: DE overlap (FDR)</a>
    </li>
    <li>
      <a href="RFresults_forPaper.html">Fig 4: RF results</a>
    </li>
    <li>
      <a href="NicheNet_forPaper.html">Fig 5: NicheNet</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Biological factors affecting transcriptional response to ligand stimulus</h1>

</div>


<p>Associating receptor mRNA / protein abundance with transcriptional response. The quantile-normalized transcriptomes per cell type (level 3 data) will be used to determine whether the cognate receptor for each ligand is present.</p>
<div id="receptor-mapping-from-ccinx" class="section level1">
<h1>Receptor mapping from CCInx</h1>
<p>Ligand-receptor interactions are from the Bader lab database included in <a href="https://baderlab.github.io/CCInx/">CCInx</a>.</p>
<div id="receptor-expression-per-cell-line" class="section level2">
<h2>Receptor expression per cell line</h2>
<pre><code>## 91% (14%) of EGF receptors present in CMap data, either inferred or (assayed). 
## 100% (8%) of HGF receptors present in CMap data, either inferred or (assayed). 
## 94% (25%) of INS receptors present in CMap data, either inferred or (assayed). 
## 100% (19%) of IGF2 receptors present in CMap data, either inferred or (assayed). 
## 100% (33%) of TGFA receptors present in CMap data, either inferred or (assayed). 
## 100% (15%) of IGF1 receptors present in CMap data, either inferred or (assayed). 
## 94% (15%) of FGF1 receptors present in CMap data, either inferred or (assayed). 
## 100% (31%) of HBEGF receptors present in CMap data, either inferred or (assayed). 
## 97% (11%) of IL6 receptors present in CMap data, either inferred or (assayed). 
## 98% (12%) of IL4 receptors present in CMap data, either inferred or (assayed). 
## 93% (9%) of IFNG receptors present in CMap data, either inferred or (assayed). 
## 100% (40%) of BTC receptors present in CMap data, either inferred or (assayed). 
## 89% (11%) of GDNF receptors present in CMap data, either inferred or (assayed). 
## 100% (0%) of GAS6 receptors present in CMap data, either inferred or (assayed). 
## 99% (12%) of TNF receptors present in CMap data, either inferred or (assayed). 
## 100% (9%) of IL17A receptors present in CMap data, either inferred or (assayed).</code></pre>
<p>Below is an overview of all potential cognate receptors per ligand present in the data. The left figure shows which ligands (dark purple) each receptor is known to interact with. On the right is a heatmap showing mean normalized gene expression per cell type for each receptor gene, with darker indicating higher expression. Gene names (rows) are ordered by heirarchical clustering of ligand-receptor interaction data, such that receptors common to the same ligand should be together. <strong>Landmark</strong> gene expression was emperically determined rather than inferred (names indicated in bold). Ligand names (columns, left) are ordered by heirarchical clustering of ligand-receptor interaction data, such that ligands sharing the same receptors should be proximal. Cell type names (columns, right) are ordered by heirarchical clustering of mean receptor gene expression.</p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_ccinx-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/rec_per_lig_ccinx-1.png" width="864" /></p>
</div>
<div id="rf-accuracy-explained-by-receptor-availability" class="section level2">
<h2>RF accuracy explained by receptor availability?</h2>
<p>Cognate receptor expression for each ligand was compared to the accuracy per cell line of the leave-one-out ligand classification model outlined <a href="https://baderlab.github.io/Brendan_CCInxPred/LigPred_lvl5.html#cell-type-biases-in-accuracy">previously</a>. Lines for each receptor are coloured by the spearman correlation coefficient between expression and accuracy per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.</p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_acc_corr_ccinx-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/legend1-1.png" width="864" /></p>
</div>
<div id="number-of-de-genes-explained-by-receptor-availability" class="section level2">
<h2>Number of DE genes explained by receptor availability?</h2>
<p><img src="RecExpr_FDR_files/figure-html/lig_DE_ccinx-1.png" width="864" /></p>
<p>Cognate receptor expression for each ligand was compared to probability scores for the number of differentially expressed genes associated with each ligand-treated cell line outlined <a href="https://baderlab.github.io/Brendan_CCInxPred/DEoverlap_FDR.html#averaging-within-each-ligand-per-cell-line">previously</a>. Lines for each receptor are coloured by the spearman correlation coefficient between expression and DE score per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.</p>
<p><img src="RecExpr_FDR_files/figure-html/rec_DE_corr_detail_ccinx-1.png" width="864" /> <img src="RecExpr_FDR_files/figure-html/legend2-1.png" width="864" /></p>
<hr />
</div>
</div>
<div id="receptor-mapping-from-fantom5" class="section level1">
<h1>Receptor mapping from FANTOM5</h1>
<p>Ligand-receptor interactions are from the FANTOM5 database (Ramilowski et al.)</p>
<div id="receptor-expression-per-cell-line-1" class="section level2">
<h2>Receptor expression per cell line</h2>
<pre><code>## 100% (50%) of EGF receptors present in CMap data, either inferred or (assayed). 
## 100% (20%) of HGF receptors present in CMap data, either inferred or (assayed). 
## 100% (33%) of INS receptors present in CMap data, either inferred or (assayed). 
## 100% (67%) of IGF2 receptors present in CMap data, either inferred or (assayed). 
## 100% (75%) of TGFA receptors present in CMap data, either inferred or (assayed). 
## 100% (50%) of IGF1 receptors present in CMap data, either inferred or (assayed). 
## 88% (50%) of FGF1 receptors present in CMap data, either inferred or (assayed). 
## 100% (43%) of HBEGF receptors present in CMap data, either inferred or (assayed). 
## 100% (0%) of IL6 receptors present in CMap data, either inferred or (assayed). 
## 100% (40%) of IL4 receptors present in CMap data, either inferred or (assayed). 
## 100% (0%) of IFNG receptors present in CMap data, either inferred or (assayed). 
## 100% (75%) of BTC receptors present in CMap data, either inferred or (assayed). 
## 80% (0%) of GDNF receptors present in CMap data, either inferred or (assayed). 
## 100% (0%) of GAS6 receptors present in CMap data, either inferred or (assayed). 
## 83% (17%) of TNF receptors present in CMap data, either inferred or (assayed). 
## 100% (0%) of IL17A receptors present in CMap data, either inferred or (assayed).</code></pre>
<p>Below is an overview of all potential cognate receptors per ligand present in the data. The left figure shows which ligands (dark purple) each receptor is known to interact with. On the right is a heatmap showing mean normalized gene expression per cell type for each receptor gene, with darker indicating higher expression. Gene names (rows) are ordered by heirarchical clustering of ligand-receptor interaction data, such that receptors common to the same ligand should be together. <strong>Landmark</strong> gene expression was emperically determined rather than inferred (names indicated in bold). Ligand names (columns, left) are ordered by heirarchical clustering of ligand-receptor interaction data, such that ligands sharing the same receptors should be proximal. Cell type names (columns, right) are ordered by heirarchical clustering of mean receptor gene expression.</p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_FANTOM5-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/rec_per_lig_FANTOM5-1.png" width="864" /></p>
</div>
<div id="rf-accuracy-explained-by-receptor-availability-1" class="section level2">
<h2>RF accuracy explained by receptor availability?</h2>
<p>Cognate receptor expression for each ligand was compared to the accuracy per cell line of the leave-one-out ligand classification model outlined <a href="https://baderlab.github.io/Brendan_CCInxPred/LigPred_lvl5.html#cell-type-biases-in-accuracy">previously</a>. Lines for each receptor are coloured by the spearman correlation coefficient between expression and accuracy per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.</p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_acc_corr_FANTOM5-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/legend3-1.png" width="864" /></p>
</div>
<div id="number-of-de-genes-explained-by-receptor-availability-1" class="section level2">
<h2>Number of DE genes explained by receptor availability?</h2>
<p><img src="RecExpr_FDR_files/figure-html/lig_DE_FANTOM5-1.png" width="864" /></p>
<p>Cognate receptor expression for each ligand was compared to probability scores for the number of differentially expressed genes associated with each ligand-treated cell line outlined <a href="https://baderlab.github.io/Brendan_CCInxPred/DEoverlap_FDR.html#averaging-within-each-ligand-per-cell-line">previously</a>. Lines for each receptor are coloured by the spearman correlation coefficient between expression and DE score per cell type (purple is negative, green is positive), and the three most correlated receptor genes are labeled.</p>
<p><img src="RecExpr_FDR_files/figure-html/rec_DE_corr_detail_FANTOM5-1.png" width="864" /></p>
</div>
<div id="transcriptional-activity-score-explained-by-receptor-availability" class="section level2">
<h2>Transcriptional Activity Score explained by receptor availability?</h2>
<p><img src="RecExpr_FDR_files/figure-html/TAS_acc_pDE_corr-1.png" width="768" /></p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_TAS_corr_FANTOM5-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/legend5-1.png" width="864" /></p>
<hr />
</div>
</div>
<div id="proteomics-from-depmap-receptor-mapping-from-fantom5" class="section level1">
<h1>Proteomics from DepMap, receptor mapping from FANTOM5</h1>
<p>Ligand-receptor interactions are from the FANTOM5 database (Ramilowski et al.)</p>
<div id="receptor-protein-abundance-per-cell-line" class="section level2">
<h2>Receptor protein abundance per cell line</h2>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_TAS_corr_depmap-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/legend6-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_acc_corr_depmap-1.png" width="864" /></p>
<p><img src="RecExpr_FDR_files/figure-html/rec_expr_DE_corr_depmap-1.png" width="864" /></p>
</div>
<div id="correlation-between-mrna-and-protein-abundance" class="section level2">
<h2>Correlation between mRNA and protein abundance</h2>
<p><img src="RecExpr_FDR_files/figure-html/expr_prot_corr-1.png" width="576" /></p>
</div>
</div>
<div id="summary-figures" class="section level1">
<h1>Summary figures</h1>
<p><img src="RecExpr_FDR_files/figure-html/summary-1.png" width="768" /></p>
<div id="ligands-in-9-cell-lines" class="section level2">
<h2>295 ligands in 9 cell lines</h2>
<p><img src="RecExpr_FDR_files/figure-html/295_expr_prot_corr-1.png" width="576" /></p>
<p><img src="RecExpr_FDR_files/figure-html/summary295-1.png" width="768" /></p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
